{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dd409128-cb77-4940-bd33-2abd652bf170",
   "metadata": {},
   "source": [
    "# Script to analyse the Orthologs that were only found by the assembly based methods fDOG-Assembly and BUSCO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bb5cc83a-d502-41a9-b2ba-fad0e75dc940",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96089966-9b43-4e6a-a6ca-dc6240d82f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "tools_ref = ['bbh', 'domainoid', 'ensemble', 'hieranoid', 'inparanoid', 'metaphors', 'oma_pairs', 'orthoffgc', 'orthofinder', 'orthoinspector', 'panther', 'rsd', 'sonicparanoid']\n",
    "#tools = ['busco_metazoa_augustus', 'busco_metazoa_metaeuk', 'fdog_assembly_metazoa_augustus_galga_v2','fdog_assembly_metazoa_augustus', 'fdog_assembly_metazoa_augustus_fly', 'fdog_assembly_metazoa_metaeuk', 'fdog_assembly_metazoa_sens_metaeuk']\n",
    "tools = ['busco_metazoa_augustus_gallus_v2', 'busco_metazoa_augustus_species_gallus_v2', 'busco_metazoa_metaeuk_gallus_v2', 'compleasm_metazoa_gallus_v2', 'fdog_assembly_metazoa_augustus_gallus_v2', 'fdog_assembly_metazoa_metaeuk_gallus_v2_sens']\n",
    "path = '../../results/qfo_input/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ddcc1b34-a91b-432c-aa8f-ee1baa54c74e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_set_of_sets(path):\n",
    "    file = open(path, 'r')\n",
    "    pairs_set = set()\n",
    "    lines = file.readlines()\n",
    "    for line in lines:\n",
    "        line = line.rstrip()\n",
    "        pairs = frozenset(line.split('\\t'))\n",
    "        pairs_set.add(pairs)\n",
    "    #print(len(lines))\n",
    "    print(len(pairs_set))\n",
    "    return pairs_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ac3d18e0-e852-41e1-b8e6-6a22c4e3e08a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8738\n",
      "8945\n",
      "8168\n",
      "8422\n",
      "8487\n",
      "8979\n",
      "7973\n",
      "8616\n",
      "9372\n",
      "9006\n",
      "8906\n",
      "8606\n",
      "9126\n"
     ]
    }
   ],
   "source": [
    "dict_of_sets = {}\n",
    "for i in tools_ref:\n",
    "    pairs_set = create_set_of_sets(path + i + '.tsv')\n",
    "    dict_of_sets[i] = pairs_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "55763023-a006-40a0-9593-e412c1a2620e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "busco_metazoa_augustus_gallus_v2\n",
      "8567\n",
      "busco_metazoa_augustus_species_gallus_v2\n",
      "8588\n",
      "busco_metazoa_metaeuk_gallus_v2\n",
      "8621\n",
      "compleasm_metazoa_gallus_v2\n",
      "8714\n",
      "fdog_assembly_metazoa_augustus_gallus_v2\n",
      "8232\n",
      "fdog_assembly_metazoa_metaeuk_gallus_v2_sens\n",
      "8384\n"
     ]
    }
   ],
   "source": [
    "for t in tools:\n",
    "    print(t)\n",
    "    pairs_set = create_set_of_sets(path + t + '.tsv')\n",
    "    dict_of_sets[t] = pairs_set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f89e2682-28d2-45e3-8ace-945ec44a5003",
   "metadata": {},
   "source": [
    "## Analyse assigned genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3578a5a9-da9a-4b69-8c0e-b812415a445f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "busco_metazoa_augustus_gallus_v2\n",
      "All pairs\n",
      "8567\n",
      "Not found by any of the protein based tools\n",
      "492\n",
      "busco_metazoa_augustus_species_gallus_v2\n",
      "All pairs\n",
      "8588\n",
      "Not found by any of the protein based tools\n",
      "354\n",
      "busco_metazoa_metaeuk_gallus_v2\n",
      "All pairs\n",
      "8621\n",
      "Not found by any of the protein based tools\n",
      "168\n",
      "compleasm_metazoa_gallus_v2\n",
      "All pairs\n",
      "8714\n",
      "Not found by any of the protein based tools\n",
      "91\n",
      "fdog_assembly_metazoa_augustus_gallus_v2\n",
      "All pairs\n",
      "8232\n",
      "Not found by any of the protein based tools\n",
      "133\n",
      "fdog_assembly_metazoa_metaeuk_gallus_v2_sens\n",
      "All pairs\n",
      "8384\n",
      "Not found by any of the protein based tools\n",
      "127\n"
     ]
    }
   ],
   "source": [
    "#get the exclusive pairs for each assembly_based tool\n",
    "exclusive_pair_dict = {}\n",
    "for t in tools:\n",
    "    exclusive_pairs = dict_of_sets[t]\n",
    "    for ref_t in tools_ref:\n",
    "        exclusive_pairs = exclusive_pairs - dict_of_sets[ref_t]\n",
    "    print(t)\n",
    "    print('All pairs')\n",
    "    print(len(dict_of_sets[t]))\n",
    "    print('Not found by any of the protein based tools')\n",
    "    print(len(exclusive_pairs))\n",
    "    exclusive_pair_dict[t] = exclusive_pairs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "83a2c2a8-abec-414e-ae7d-fa32a38c1f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get seed genes\n",
    "seed_file = open('../uniprotid_to_group_assignment/mapping_busco_id_uniport_id.tsv', 'r')\n",
    "lines = seed_file.readlines()\n",
    "seed_genes = set()\n",
    "busco_vs_uniprot_dict = {}\n",
    "for line in lines:\n",
    "    line = line.rstrip()\n",
    "    busco_id, uniprot_id = line.split('\\t')\n",
    "    seed_genes.add(uniprot_id)\n",
    "    busco_vs_uniprot_dict[busco_id] = uniprot_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "20ec1274-034b-42db-8446-8d0249ab28b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_species_file(file):\n",
    "    lines = file.readlines()\n",
    "    species_dict = {}\n",
    "    for line in lines:\n",
    "        line = line.rstrip()\n",
    "        ncbi, name, uniprot_acc, source, refseq_acc = line.split('\\t')\n",
    "        species_dict[ncbi] = {'name': name, 'uniprot': uniprot_acc, 'source': source, 'refseq': refseq_acc}\n",
    "    return species_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "baad0f5f-91bd-4ac4-994a-69c5f62e4ecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniprot_ids_for_species(file_path):\n",
    "    file = open(file_path, 'r')\n",
    "    lines = file.readlines()\n",
    "    id_set = set()\n",
    "    for line in lines:\n",
    "        line = line.rstrip()\n",
    "        uniprot_id = line.split('\\t')[0]\n",
    "        id_set.add(uniprot_id)\n",
    "    file.close()\n",
    "    return id_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a8b9904b-9d9d-4791-95bd-c3c12cac5ed6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nematostella vectensis\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000001593_45351.idmapping\n",
      "Rattus norvegicus\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000002494_10116.idmapping\n",
      "Gallus gallus\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000000539_9031.idmapping\n",
      "Xenopus tropicalis\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000008143_8364.idmapping\n",
      "Danio rerio\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000000437_7955.idmapping\n",
      "Drosophila melanogaster\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000000803_7227.idmapping\n",
      "Tribolium castaneum\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000007266_7070.idmapping\n",
      "Ixodes scapularis\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000001555_6945.idmapping\n",
      "Helobdella robusta\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000015101_6412.idmapping\n",
      "Caenorhabditis elegans\n",
      "../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/UP000001940_6239.idmapping\n"
     ]
    }
   ],
   "source": [
    "#create species to uniprot mapping\n",
    "species_file = open('../../data/fDOG-assembly/species_set_benchmark_v2.tsv', 'r')\n",
    "species_dict = parse_species_file(species_file)\n",
    "species_file.close()\n",
    "\n",
    "species_list = []\n",
    "uniprot_dict = {} # dictionary species_ncbi_id:set(uniprot_ids of species)\n",
    "for species in species_dict:\n",
    "    print(species_dict[species]['name'])\n",
    "    path = '../../data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/' + species_dict[species]['uniprot'] + '_' + species + '.idmapping'\n",
    "    print(path)\n",
    "    uniprot_dict[species] = get_uniprot_ids_for_species(path)\n",
    "    species_list.append(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5e67962f-c041-4f19-9667-29fc7843071b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_species(species_dict, gene):\n",
    "    for sp in species_dict:\n",
    "        if gene in species_dict[sp]:\n",
    "            return sp\n",
    "    print(gene)\n",
    "    return 'Error'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ee3a7a15-ec07-4f96-8d17-4318468d5e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create matrix seed gene vs species and count how many tools found at least one pair\n",
    "# blank matrix seed vs species\n",
    "df_ref_tools = pd.DataFrame(0, index=list(seed_genes), columns=species_list) \n",
    "#print(df_ref_tools)\n",
    "seed_dict = {}\n",
    "\n",
    "#fill matrix\n",
    "for tool in tools_ref:\n",
    "    #print(tool)\n",
    "    for pair in dict_of_sets[tool]:\n",
    "        gene1, gene2 = pair\n",
    "        if gene1 in seed_genes:\n",
    "            seed = gene1\n",
    "            ortholog = gene2\n",
    "        else:\n",
    "            ortholog = gene1\n",
    "            seed = gene2\n",
    "        species = assign_species(uniprot_dict, ortholog)\n",
    "        if species == 'Error':\n",
    "            print('Ortholog gene could not be assigned to species')\n",
    "            print(ortholog)\n",
    "            break\n",
    "        try:\n",
    "            seed_dict[seed][species].add(tool)\n",
    "        except KeyError:\n",
    "            try:\n",
    "                seed_dict[seed][species] = set()\n",
    "                seed_dict[seed][species].add(tool)\n",
    "            except KeyError:    \n",
    "                seed_dict[seed] = {}\n",
    "                seed_dict[seed][species] = set()\n",
    "                seed_dict[seed][species].add(tool)\n",
    "\n",
    "for seed in seed_dict:\n",
    "    for species in seed_dict[seed]:\n",
    "        no_tools = len(seed_dict[seed][species])\n",
    "        df_ref_tools.loc[seed, species] = no_tools\n",
    "\n",
    "#print(df_ref_tools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "42b25d90-d8ff-457e-9014-34709f6333eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_exclusive_genes_assigned_to_uniprot(exclusive_pairs, seed_genes, uniprot_dict, df_ref_tools):\n",
    "    # genes that fill a gap in the profile -> turn 0 in 1\n",
    "    fills_gap = 0\n",
    "    # genes found in all the other species\n",
    "    conserved = 0\n",
    "    # genes other tools found an ortholog but another one \n",
    "    no_gap = 0\n",
    "    # the protein-based tools don't include self-assignments \n",
    "    self_assignment = 0\n",
    "    species_counter = {}\n",
    "\n",
    "    filled_gaps = set()\n",
    "    conserved_filled_gaps = set()\n",
    "    no_gaps = set()\n",
    "    \n",
    "    for pair in exclusive_pairs:\n",
    "        try:\n",
    "            gene1, gene2 = pair\n",
    "        except ValueError:\n",
    "            #print(pair)\n",
    "            self_assignment += 1\n",
    "            gene1 = list(pair)[0]\n",
    "            gene2 = gene1\n",
    "        #assign genes to seed and ortholog\n",
    "        if gene1 in seed_genes:\n",
    "            seed = gene1\n",
    "            ortholog = gene2\n",
    "        else:\n",
    "            seed = gene2\n",
    "            ortholog = gene1\n",
    "        #assign ortholog to species\n",
    "        if species == 'Error':\n",
    "            print('Ortholog gene could not be assigned to species')\n",
    "            print(ortholog)\n",
    "            #break\n",
    "                \n",
    "        if df_ref_tools.loc[seed, species] == 0:\n",
    "            fills_gap += 1\n",
    "            filled_gaps.add(str(species) + '_' + seed)\n",
    "            #check conservation\n",
    "            zeros = (df_ref_tools.loc[seed] == 0).sum()\n",
    "            if zeros <= 1:\n",
    "                conserved += 1\n",
    "                if str(species) + '_' + seed in conserved_filled_gaps:\n",
    "                    continue\n",
    "                else:\n",
    "                    conserved_filled_gaps.add(str(species) + '_' + seed)\n",
    "                    try:\n",
    "                        species_counter[species] += 1\n",
    "                    except KeyError:\n",
    "                        species_counter[species] = 1\n",
    "                \n",
    "            #else:\n",
    "             #   print(df_ref_tools.loc[seed])\n",
    "        else:\n",
    "            no_gap += 1\n",
    "            no_gaps.add(str(species) + '_' + seed)\n",
    "            \n",
    "    return fills_gap, len(filled_gaps), conserved, len(conserved_filled_gaps), no_gap, len(no_gaps), self_assignment, species_counter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "542d29b4-ca7b-4e94-b8b9-9048a6bcbb58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "busco_metazoa_augustus_gallus_v2\n",
      "492\n",
      "Pairs that filled a gap, Filled gaps in Profile, Pairs that filled a gap of a conserved gene, Conserved gaps filled, Pairs that did not fill a gap, On gene level, Self-assignment\n",
      "(31, 26, 26, 21, 461, 356, 13, {'6945': 21})\n",
      "busco_metazoa_augustus_species_gallus_v2\n",
      "354\n",
      "Pairs that filled a gap, Filled gaps in Profile, Pairs that filled a gap of a conserved gene, Conserved gaps filled, Pairs that did not fill a gap, On gene level, Self-assignment\n",
      "(16, 14, 14, 12, 338, 271, 13, {'6945': 12})\n",
      "busco_metazoa_metaeuk_gallus_v2\n",
      "168\n",
      "Pairs that filled a gap, Filled gaps in Profile, Pairs that filled a gap of a conserved gene, Conserved gaps filled, Pairs that did not fill a gap, On gene level, Self-assignment\n",
      "(12, 10, 11, 9, 156, 133, 14, {'6945': 9})\n",
      "compleasm_metazoa_gallus_v2\n",
      "91\n",
      "Pairs that filled a gap, Filled gaps in Profile, Pairs that filled a gap of a conserved gene, Conserved gaps filled, Pairs that did not fill a gap, On gene level, Self-assignment\n",
      "(5, 5, 5, 5, 86, 82, 14, {'6945': 5})\n",
      "fdog_assembly_metazoa_augustus_gallus_v2\n",
      "133\n",
      "Pairs that filled a gap, Filled gaps in Profile, Pairs that filled a gap of a conserved gene, Conserved gaps filled, Pairs that did not fill a gap, On gene level, Self-assignment\n",
      "(5, 5, 5, 5, 128, 119, 13, {'6945': 5})\n",
      "fdog_assembly_metazoa_metaeuk_gallus_v2_sens\n",
      "127\n",
      "Pairs that filled a gap, Filled gaps in Profile, Pairs that filled a gap of a conserved gene, Conserved gaps filled, Pairs that did not fill a gap, On gene level, Self-assignment\n",
      "(2, 2, 2, 2, 125, 115, 13, {'6945': 2})\n"
     ]
    }
   ],
   "source": [
    "for t in tools:\n",
    "    print(t)\n",
    "    print(len(exclusive_pair_dict[t]))\n",
    "    print(\"Pairs that filled a gap, Filled gaps in Profile, Pairs that filled a gap of a conserved gene, Conserved gaps filled, Pairs that did not fill a gap, On gene level, Self-assignment\")\n",
    "    print(evaluate_exclusive_genes_assigned_to_uniprot(exclusive_pair_dict[t], seed_genes, uniprot_dict, df_ref_tools))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "999c9a11-42e6-441a-aa41-331a834526e5",
   "metadata": {},
   "source": [
    "## Analyse gap distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2ed5666d-432f-4f91-8035-4b8751000582",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45351    33\n",
      "10116     7\n",
      "9031     40\n",
      "8364     18\n",
      "7955      1\n",
      "7227     19\n",
      "7070      5\n",
      "6945     52\n",
      "6412     36\n",
      "6239     80\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# how many gaps are there in total in the df_ref_tools?\n",
    "zeros = (df_ref_tools == 0).sum()\n",
    "print(zeros)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e7278f49-9808-40d7-be4f-5b2056c1ff0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "214\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>45351</th>\n",
       "      <th>10116</th>\n",
       "      <th>9031</th>\n",
       "      <th>8364</th>\n",
       "      <th>7955</th>\n",
       "      <th>7227</th>\n",
       "      <th>7070</th>\n",
       "      <th>6945</th>\n",
       "      <th>6412</th>\n",
       "      <th>6239</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P22570</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q16795</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y3D8</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q8NI36</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q15061</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9GZP4</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q8TAM2</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y276</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q96MW1</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q8N5M1</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>917 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        45351  10116  9031  8364  7955  7227  7070  6945  6412  6239\n",
       "P22570     12     12    13    13    13    13    13    12    13    12\n",
       "Q16795     13     13    13    13    13    12    13    11    13    12\n",
       "Q9Y3D8     12     12    12    12    12    12     0    11    12    12\n",
       "Q8NI36      0     13    12    12    12    12    12    12    12    11\n",
       "Q15061     13     13    13    12    13    13    11    12    12     7\n",
       "...       ...    ...   ...   ...   ...   ...   ...   ...   ...   ...\n",
       "Q9GZP4     13     13    13    12    13    13    13    12    13    13\n",
       "Q8TAM2     13     13    13    13    13    13    13     6    13    13\n",
       "Q9Y276     13     13    13    13    13    13    13    12    13    13\n",
       "Q96MW1      0     13    13    13    12    13    13     6     7     0\n",
       "Q8N5M1     13     13    13    13    13    13    13    11    13    13\n",
       "\n",
       "[917 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# how many of them are conserved?\n",
    "count_exactly_one_zero = (df_ref_tools == 0).sum(axis=1).eq(1).sum()\n",
    "print(count_exactly_one_zero)\n",
    "display(df_ref_tools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3459f099-be5f-48ab-8f34-d2e762c1217f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count of columns where the single zeros are located:\n",
      "45351    26\n",
      "10116     5\n",
      "9031     32\n",
      "8364     14\n",
      "7955      1\n",
      "7227     13\n",
      "7070      4\n",
      "6945     43\n",
      "6412     21\n",
      "6239     55\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Identify rows with exactly one 0\n",
    "rows_with_one_zero = (df_ref_tools == 0).sum(axis=1).eq(1)\n",
    "\n",
    "# Filter the DataFrame to include only rows with exactly one 0\n",
    "filtered_df = df_ref_tools[rows_with_one_zero]\n",
    "\n",
    "# Count occurrences of 0 in each column within the filtered rows\n",
    "column_zero_counts = (filtered_df == 0).sum()\n",
    "\n",
    "print(\"Count of columns where the single zeros are located:\")\n",
    "print(column_zero_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "71faeced-7101-4e94-9c86-6dbd83817bfe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        45351  10116  9031  8364  7955  7227  7070  6945  6412  6239\n",
      "Q9BSF4     13      0     0    13    13    12     0     2     0    13\n",
      "Q96MX6     13      0    13    13    13    13    13     7    13     1\n",
      "Q8TDD1     13      0    13    12    13    13    13    12    12    13\n",
      "O75419     13      0    13    13    13    13    13    12    13    13\n",
      "P61758     13      0    13    13    13    13    13    12    12    13\n",
      "Q96AB6      8      0    13    13    11    13    13    12    13     0\n",
      "O43462     13      0    13    13    13    13    13    12    13    13\n"
     ]
    }
   ],
   "source": [
    "# Has it worked correctly?\n",
    "print(df_ref_tools[df_ref_tools['10116'] == 0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4aca1ec-8f1f-4c99-ae55-e5177954df72",
   "metadata": {},
   "source": [
    "## Analyse unassigned genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "752299cc-0d13-4b93-a483-d499ddddecbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get genes that were not assigned to a UniProt ID\n",
    "# for that we have to parse the overlap tables and filter out the genes that have no overlap at all with a CDS in the reference gff files or that we could not assign to a UniProt gene\n",
    "busco_augustus_df = pd.read_csv('../overlap_tables/busco_augustus_overlap_gff_files_gallus_v2.tsv', delimiter='\\t')\n",
    "busco_augustus_species_df = pd.read_csv('../overlap_tables/busco_augustus_species_overlap_gff_files_gallus_v2.tsv', delimiter='\\t')\n",
    "busco_metaeuk_df = pd.read_csv('../overlap_tables/busco_metaeuk_overlap_gff_files_gallus_2.tsv', delimiter='\\t')\n",
    "fa_augustus_df = pd.read_csv('../overlap_tables/fdog_ass_busco_augustus_overlap_gff_files_gallus_v2.tsv', delimiter='\\t')\n",
    "fa_metaeuk_df = pd.read_csv('../overlap_tables/fdog_ass_busco_metaeuk_overlap_gff_files_gallus_v2_sens.tsv', delimiter='\\t')\n",
    "compleasm_df = pd.read_csv('../overlap_tables/compleasm_overlap_gff_files_gallus_v2.tsv', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9f1a08ac-3a1b-421c-91da-7648c25e4be9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_unassigned_genes(exclusive_genes, df_ref_tools):\n",
    "    # genes that fill a gap in the profile -> turn 0 in 1\n",
    "    fills_gap = 0\n",
    "    # genes found in all the other species\n",
    "    conserved = 0\n",
    "    # genes other tools found an ortholog but another one \n",
    "    no_gap = 0\n",
    "    # the protein-based tools don't include self-assignments \n",
    "    self_assignment = 0\n",
    "\n",
    "    species_counter = {}\n",
    "\n",
    "    filled_gaps = set()\n",
    "    conserved_filled_gaps = set()\n",
    "    no_gaps = set()\n",
    "\n",
    "    for pair in exclusive_genes:\n",
    "        species, seed = pair\n",
    "        if df_ref_tools.loc[seed, str(species)] == 0:\n",
    "            fills_gap += 1\n",
    "            filled_gaps.add(str(species) + '_' + seed)\n",
    "            if species == 10116:\n",
    "                print('Seed')\n",
    "                print(seed)\n",
    "            #check conservation\n",
    "            zeros = (df_ref_tools.loc[seed] == 0).sum()\n",
    "            if zeros <= 1:\n",
    "                conserved += 1\n",
    "                if str(species) + '_' + seed in conserved_filled_gaps:\n",
    "                    continue\n",
    "                else:\n",
    "                    conserved_filled_gaps.add(str(species) + '_' + seed)\n",
    "                    try:\n",
    "                        species_counter[species] += 1\n",
    "                    except KeyError:\n",
    "                        species_counter[species] = 1\n",
    "            #else:\n",
    "             #   print(df_ref_tools.loc[seed])\n",
    "        else:\n",
    "            no_gap += 1\n",
    "            no_gaps.add(str(species) + '_' + seed)\n",
    "        \n",
    "\n",
    "    return fills_gap, len(filled_gaps), conserved, len(conserved_filled_gaps), no_gap, len(no_gaps), self_assignment, species_counter, conserved_filled_gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c67b071e-84f0-4756-903f-e4afb22bd840",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unassigned_exclusive_pairs(df, busco_to_uniprot):\n",
    "    unassigned = df[df.isnull().any(axis=1)]\n",
    "    #print(unassigned)\n",
    "    pairs_rows = unassigned[['Species','GeneID']].copy()\n",
    "    pairs_rows['GeneID'] = pairs_rows['GeneID'].map(busco_to_uniprot)\n",
    "    pairs_rows.dropna(inplace = True)\n",
    "    #print(pairs_rows)\n",
    "    print(pairs_rows['Species'].value_counts())\n",
    "    #unassigned_pairs = pairs_rows.apply(frozenset, axis=1)\n",
    "    unassigned_pairs = pairs_rows.values.tolist()\n",
    "    return unassigned_pairs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1338904f-85df-4f4f-8ffd-8fac0c618f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_seed_genes(unassigned_pairs):\n",
    "    seeds = set()\n",
    "    pairs = set()\n",
    "    for i in unassigned_pairs:\n",
    "        sp, seed = i\n",
    "        seeds.add(seed)\n",
    "        pairs.add(str(sp) + '_' + seed)\n",
    "    return len(seeds), len(pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "410d5706-232f-4e44-9511-4a518fb72527",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6945     113\n",
      "45351     71\n",
      "8364      57\n",
      "10116     51\n",
      "9031      38\n",
      "6412      25\n",
      "6239      12\n",
      "7227      11\n",
      "7070       6\n",
      "7955       6\n",
      "Name: Species, dtype: int64\n",
      "390\n",
      "(286, 332)\n",
      "################\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "Q96AB6\n",
      "(72, 68, 55, 53, 318, 264, 0, {9031: 4, 10116: 1, 7070: 2, 6239: 1, 45351: 14, 6945: 23, 7955: 1, 6412: 5, 8364: 2}, {'6945_Q9NRN9', '45351_Q14CX7', '6945_Q8N5C7', '6945_P62277', '45351_Q8IWT0', '6945_Q9BS26', '7070_P57081', '45351_Q96A65', '6945_Q5T280', '6945_P62829', '6412_Q9Y3D3', '6945_Q9UJK0', '6945_Q9P2I0', '8364_Q9UK45', '45351_Q9NXG2', '6945_Q92542', '45351_Q969N2', '9031_Q9BRT9', '7070_Q9Y3D8', '6945_Q8N6T3', '6945_Q9H4B6', '10116_Q96MX6', '45351_Q9UBL3', '45351_Q92845', '6412_P40938', '9031_Q13405', '45351_Q9NQ89', '45351_Q9P1Q0', '45351_P53990', '6945_Q05048', '6945_Q9UJX6', '6945_P48553', '6945_Q9H3H5', '6239_Q9ULC3', '6945_P49959', '6945_Q330K2', '6945_Q9BXS1', '9031_Q96L58', '45351_P61221', '45351_Q9BPX7', '6412_O75380', '8364_Q9Y2S7', '6945_Q9Y248', '9031_Q9HCU5', '6945_O94817', '6412_Q9BPZ7', '45351_P12004', '7955_O75165', '6412_Q32P41', '6945_Q07864', '6945_O14653', '45351_Q9NVU0', '6945_Q8TAT6'})\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "Q96AB6\n"
     ]
    }
   ],
   "source": [
    "#fDOG-Assembly Augustus\n",
    "unassigned_pairs = get_unassigned_exclusive_pairs(fa_augustus_df, busco_vs_uniprot_dict)\n",
    "#how many pairs are included?\n",
    "print(len(unassigned_pairs))\n",
    "# how many different seed genes are included ?\n",
    "print(number_seed_genes(unassigned_pairs))\n",
    "print('################')\n",
    "print(evaluate_unassigned_genes(unassigned_pairs, df_ref_tools))\n",
    "conserved_gaps_fda_aug = evaluate_unassigned_genes(unassigned_pairs, df_ref_tools)[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cc003462-01c1-46de-bd4f-a12fada3cb17",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unexpected indent (2803421596.py, line 4)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipykernel_2230572/2803421596.py\"\u001b[0;36m, line \u001b[0;32m4\u001b[0m\n\u001b[0;31m    147873at33208_CM026976_1_1_g3.t1\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unexpected indent\n"
     ]
    }
   ],
   "source": [
    "# for some additional investigations\n",
    "#unassigned_pairs[species].value_counts()\n",
    "#missing in overlap table:  why? -> fixed, due to None as UniProt IDs -> error by grouping the df during overlap table reconstruction\n",
    "    147873at33208_CM026976_1_1_g3.t1\n",
    "\t325183at33208_CM026974_1_1_g2.t1\n",
    "\t488635at33208_CM026978_1_1_g3.t1\n",
    "\t491289at33208_CM026989_1_1_g2.t1\n",
    "\t526673at33208_CM026983_1_1_g1.t1\n",
    "\t535367at33208_CM026984_1_1_g2.t1\n",
    "\t539043at33208_CM026975_1_1_g2.t1\n",
    "\t557834at33208_CM026985_1_1_g1.t1\n",
    "\t577339at33208_CM026975_1_1_g1.t1\n",
    "\t599320at33208_CM026979_1_1_g1.t1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "02c96f6c-df6e-4f49-a7a1-313d533a0808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6945     189\n",
      "45351    120\n",
      "10116    112\n",
      "8364      88\n",
      "9031      69\n",
      "6412      26\n",
      "7227      23\n",
      "6239      19\n",
      "7070      15\n",
      "7955      10\n",
      "Name: Species, dtype: int64\n",
      "671\n",
      "Seed\n",
      "Q9BSF4\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "P61758\n",
      "(102, 91, 76, 69, 569, 466, 0, {7070: 2, 45351: 17, 9031: 6, 6412: 7, 8364: 1, 10116: 2, 6945: 34}, {'45351_Q92665', '6945_Q16514', '6945_Q9NRN9', '45351_Q14CX7', '6945_P62277', '6945_Q96K21', '6945_Q8N5C7', '45351_Q8IWT0', '6945_Q9BS26', '7070_P57081', '6945_Q9Y315', '45351_Q96A65', '6945_Q5T280', '10116_P61758', '6945_P62829', '6412_Q9Y3D3', '6945_Q9NX55', '6945_Q9UJK0', '6945_Q9P2I0', '8364_Q9UK45', '6945_Q9Y4Y9', '45351_Q9NXG2', '6945_O00623', '6945_Q92542', '45351_Q969N2', '9031_Q9BRT9', '7070_Q9Y3D8', '6945_Q9H4B6', '6945_Q8N6T3', '10116_Q96MX6', '6945_Q9NWB7', '45351_Q9UBL3', '45351_Q92845', '6412_P40938', '45351_Q9NQ89', '45351_Q9P1Q0', '9031_Q13405', '45351_P53990', '6412_Q96J42', '6945_Q05048', '6945_P48553', '6945_Q9UJX6', '6412_Q9BV87', '6945_Q9H3H5', '6945_Q9BU89', '6945_P49959', '6945_Q330K2', '6945_Q9BXS1', '9031_Q96L58', '45351_P61221', '45351_Q9BPX7', '9031_Q96EK9', '6412_O75380', '6945_Q9Y2Q9', '45351_Q8NI36', '6945_Q9NX38', '6945_Q9Y248', '9031_Q9HCU5', '6945_O94817', '6412_Q9BPZ7', '45351_P12004', '9031_Q9BT30', '6412_Q32P41', '6945_Q07864', '6945_O14653', '45351_Q9NVU0', '6945_Q9NX58', '45351_Q9NQ50', '6945_Q8TAT6'})\n",
      "Seed\n",
      "Q9BSF4\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "P61758\n"
     ]
    }
   ],
   "source": [
    "#fDOG-Assembly MetaEuk\n",
    "unassigned_pairs = get_unassigned_exclusive_pairs(fa_metaeuk_df, busco_vs_uniprot_dict)\n",
    "#print(unassigned_pairs)\n",
    "print(len(unassigned_pairs))\n",
    "print(evaluate_unassigned_genes(unassigned_pairs, df_ref_tools))\n",
    "conserved_gaps_fda_meta = evaluate_unassigned_genes(unassigned_pairs, df_ref_tools)[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3e68cccb-bb8a-48d5-80d2-9ba4d9705f79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6945     64\n",
      "10116    48\n",
      "7955     39\n",
      "45351    36\n",
      "6239     24\n",
      "6412     18\n",
      "9031      5\n",
      "7070      3\n",
      "8364      1\n",
      "Name: Species, dtype: int64\n",
      "238\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n",
      "(57, 56, 45, 44, 181, 171, 0, {9031: 1, 6412: 8, 6945: 24, 45351: 8, 10116: 1, 7070: 2}, {'6945_Q9H6L2', '45351_Q92665', '6945_Q9NRN9', '6945_P62277', '6945_Q8N5C7', '6412_Q7L592', '7070_P57081', '6945_Q5T280', '6945_Q9BUB7', '6412_A8K0Z3', '6412_Q9Y3D3', '6945_Q96NB3', '6945_Q9UJK0', '6412_Q969H6', '45351_Q9NXG2', '6945_Q92542', '9031_Q9BRT9', '6945_Q8N6T3', '6945_Q9H4B6', '7070_Q9Y3D8', '10116_Q96MX6', '6412_P40938', '45351_P53990', '6945_Q05048', '6945_Q9UJX6', '6945_P48553', '6412_Q9BV87', '6945_Q9H3H5', '6945_Q9BU89', '6945_P49959', '6945_Q9H270', '6412_O75380', '45351_Q9BPX7', '45351_P61221', '45351_P49458', '6945_Q9NX38', '6945_Q9Y248', '6945_O94817', '6412_Q32P41', '45351_P12004', '6945_Q07864', '6945_O14653', '45351_Q9NQ50', '6945_Q8TAT6'})\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n"
     ]
    }
   ],
   "source": [
    "#BUSCO Augustus\n",
    "unassigned_pairs = get_unassigned_exclusive_pairs(busco_augustus_df, busco_vs_uniprot_dict)\n",
    "#print(unassigned_pairs)\n",
    "print(len(unassigned_pairs))\n",
    "print(evaluate_unassigned_genes(unassigned_pairs, df_ref_tools))\n",
    "conserved_gaps_busco_aug = evaluate_unassigned_genes(unassigned_pairs, df_ref_tools)[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dd93835b-2a40-452b-a427-184c3ef90f22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6945     78\n",
      "45351    69\n",
      "10116    66\n",
      "8364     54\n",
      "9031     35\n",
      "7955     16\n",
      "6412     16\n",
      "6239     12\n",
      "7227      8\n",
      "7070      3\n",
      "Name: Species, dtype: int64\n",
      "357\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n",
      "(81, 78, 60, 59, 276, 236, 0, {9031: 4, 6412: 8, 6945: 24, 45351: 18, 10116: 1, 7070: 2, 8364: 2}, {'6945_Q9H6L2', '45351_Q92665', '6945_Q9NRN9', '6945_P62277', '6945_Q8N5C7', '45351_Q14CX7', '45351_Q8IWT0', '7070_P57081', '45351_Q96A65', '6945_Q5T280', '6945_Q9BUB7', '6412_A8K0Z3', '6412_Q9Y3D3', '6945_Q96NB3', '6945_Q9UJK0', '6412_Q969H6', '8364_Q9UK45', '45351_Q9NXG2', '6945_Q92542', '45351_Q969N2', '9031_Q9BRT9', '6945_Q8N6T3', '6945_Q9H4B6', '7070_Q9Y3D8', '10116_Q96MX6', '45351_Q9UBL3', '6412_P40938', '45351_Q92845', '9031_Q13405', '45351_Q9NQ89', '45351_Q9P1Q0', '45351_P53990', '6945_Q05048', '6945_Q9UJX6', '6945_P48553', '6412_Q9BV87', '6945_Q9H3H5', '6945_Q9BU89', '6945_P49959', '6412_Q8IWA0', '6945_Q9H270', '9031_Q96L58', '6412_O75380', '45351_P61221', '45351_Q9BPX7', '45351_Q8NI36', '45351_P49458', '8364_Q9Y2S7', '6945_Q9NX38', '6945_Q9Y248', '9031_Q9HCU5', '6945_O94817', '6412_Q32P41', '45351_P12004', '6945_Q07864', '6945_O14653', '45351_Q9NVU0', '45351_Q9NQ50', '6945_Q8TAT6'})\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n"
     ]
    }
   ],
   "source": [
    "#BUSCO Augustus Species\n",
    "unassigned_pairs = get_unassigned_exclusive_pairs(busco_augustus_species_df, busco_vs_uniprot_dict)\n",
    "#print(unassigned_pairs)\n",
    "print(len(unassigned_pairs))\n",
    "print(evaluate_unassigned_genes(unassigned_pairs, df_ref_tools))\n",
    "conserved_gaps_busco_aug_sp = evaluate_unassigned_genes(unassigned_pairs, df_ref_tools)[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a8019b88-fe4b-4422-bd83-c3b00270da15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10116    160\n",
      "6945      64\n",
      "6412      63\n",
      "45351     20\n",
      "9031       8\n",
      "7070       4\n",
      "8364       4\n",
      "7955       2\n",
      "6239       1\n",
      "Name: Species, dtype: int64\n",
      "326\n",
      "Seed\n",
      "Q9BSF4\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "P61758\n",
      "(74, 74, 56, 56, 252, 117, 0, {9031: 2, 6412: 11, 6945: 32, 45351: 8, 10116: 2, 7070: 1}, {'45351_Q92665', '6945_Q16514', '6945_Q9NRN9', '6945_P62277', '6945_Q96K21', '6945_Q8N5C7', '6945_Q9BS26', '6412_Q7L592', '6945_Q9Y315', '7070_P57081', '6945_Q5T280', '10116_P61758', '6945_Q9BUB7', '6412_A8K0Z3', '6945_P62829', '6412_Q9Y3D3', '6945_Q96NB3', '6945_Q9UJK0', '6945_Q9P2I0', '6412_Q969H6', '6945_O00623', '45351_Q9NXG2', '6945_Q92542', '9031_Q9BRT9', '6945_Q9H4B6', '6945_Q8N6T3', '10116_Q96MX6', '6945_Q9NWB7', '6412_P40938', '45351_P53990', '6412_Q96J42', '6945_Q05048', '6945_P48553', '6945_Q9UJX6', '6412_Q9BV87', '6945_Q99547', '6945_Q9H3H5', '6945_Q9BU89', '6945_P49959', '6945_Q330K2', '6945_Q9H270', '6412_O75380', '45351_P61221', '9031_Q96EK9', '45351_Q9BPX7', '45351_P49458', '6945_Q9Y248', '6412_Q9BPZ7', '6412_Q8N183', '6412_Q32P41', '6945_Q9NX58', '6945_Q07864', '6945_O14653', '45351_P12004', '45351_Q9NQ50', '6945_Q8TAT6'})\n",
      "Seed\n",
      "Q9BSF4\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "P61758\n"
     ]
    }
   ],
   "source": [
    "#BUSCO MetaEuk\n",
    "unassigned_pairs = get_unassigned_exclusive_pairs(busco_metaeuk_df, busco_vs_uniprot_dict)\n",
    "#print(unassigned_pairs)\n",
    "print(len(unassigned_pairs))\n",
    "print(evaluate_unassigned_genes(unassigned_pairs, df_ref_tools))\n",
    "conserved_gaps_busco_meta = evaluate_unassigned_genes(unassigned_pairs, df_ref_tools)[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7ac94d61-3553-4824-96e4-ba8ef9811ae4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6945     81\n",
      "45351    65\n",
      "8364     52\n",
      "10116    42\n",
      "9031     39\n",
      "6412     20\n",
      "7227      8\n",
      "6239      3\n",
      "7070      3\n",
      "7955      2\n",
      "Name: Species, dtype: int64\n",
      "315\n",
      "Seed\n",
      "P61758\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "Q9BSF4\n",
      "(102, 98, 77, 75, 213, 181, 0, {9031: 7, 6412: 10, 6945: 33, 45351: 19, 10116: 2, 7070: 2, 8364: 2}, {'45351_Q92665', '6945_Q16514', '6945_Q9NRN9', '6945_P62277', '6945_Q8N5C7', '45351_Q14CX7', '45351_Q8IWT0', '6945_Q9BS26', '6412_Q7L592', '6945_Q9Y315', '7070_P57081', '45351_Q96A65', '6945_Q5T280', '10116_P61758', '6412_A8K0Z3', '6945_P62829', '6412_Q9Y3D3', '6945_Q96NB3', '6945_Q9UJK0', '6945_Q9P2I0', '6412_Q969H6', '8364_Q9UK45', '6945_Q9Y4Y9', '6945_O00623', '45351_Q9NXG2', '6945_Q92542', '45351_Q969N2', '45351_Q9UNX4', '9031_Q9BRT9', '6945_Q8N6T3', '6945_Q9H4B6', '7070_Q9Y3D8', '10116_Q96MX6', '6945_Q9NWB7', '45351_Q9UBL3', '6412_P40938', '45351_Q92845', '9031_Q13405', '45351_Q9NQ89', '45351_Q9P1Q0', '45351_P53990', '6412_Q96J42', '6945_Q05048', '9031_Q8TDP1', '6945_P48553', '6945_Q9UJX6', '6412_Q9BV87', '6945_Q9H3H5', '6945_Q99547', '6945_Q9BU89', '6945_P49959', '6945_Q330K2', '6945_Q9H270', '9031_Q96L58', '6412_O75380', '9031_Q96EK9', '45351_P61221', '6945_Q9Y2Q9', '45351_Q9BPX7', '45351_Q8NI36', '45351_P49458', '8364_Q9Y2S7', '6945_Q9NX38', '6945_Q9Y248', '9031_Q9HCU5', '6412_Q9BPZ7', '9031_Q9BT30', '6412_Q32P41', '6945_Q9NX58', '6945_Q07864', '6945_O14653', '45351_Q9NVU0', '45351_P12004', '45351_Q9NQ50', '6945_Q8TAT6'})\n",
      "Seed\n",
      "P61758\n",
      "Seed\n",
      "Q96AB6\n",
      "Seed\n",
      "Q96MX6\n",
      "Seed\n",
      "Q9BSF4\n"
     ]
    }
   ],
   "source": [
    "#compleasm \n",
    "#BUSCO Augustus\n",
    "unassigned_pairs = get_unassigned_exclusive_pairs(compleasm_df, busco_vs_uniprot_dict)\n",
    "#print(unassigned_pairs)\n",
    "print(len(unassigned_pairs))\n",
    "print(evaluate_unassigned_genes(unassigned_pairs, df_ref_tools))\n",
    "conserved_gaps_busco_aug = evaluate_unassigned_genes(unassigned_pairs, df_ref_tools)[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "acdd258e-c5b9-4186-bf43-1f86532d4381",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note for my self:\n",
    "#I did not consider if I only found an additional co-ortholog or if I found additional orthologs and none of the found ones were supported by my results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "009b009f-1bc4-47d2-8234-99c4509a9397",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do the results of filled gaps by the different methods overlap?\n",
    "union_all = conserved_gaps_fda_aug.union(conserved_gaps_busco_aug, conserved_gaps_fda_aug, conserved_gaps_busco_meta)\n",
    "species = []\n",
    "uniprot = []\n",
    "count = []\n",
    "list_of_sets = [conserved_gaps_fda_aug, conserved_gaps_fda_meta, conserved_gaps_busco_aug, conserved_gaps_busco_meta]\n",
    "for i in union_all:\n",
    "    sp, uniprot_id = i.split('_')\n",
    "    count_val = 0\n",
    "    for t in list_of_sets:\n",
    "        if i in t:\n",
    "            count_val += 1\n",
    "    species.append(sp)\n",
    "    uniprot.append(uniprot_id)\n",
    "    count.append(count_val)\n",
    "\n",
    "zipped = list(zip(species, uniprot, count))\n",
    "\n",
    "df = pd.DataFrame(zipped, columns=['Species', 'UniProtID', 'Found by x tools'])\n",
    "df.to_csv('conserved_gaps_filled_by_assembly_based_methods.tsv', sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1f1558e6-2f11-4734-9333-f6fca0093ffe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>UniProtID</th>\n",
       "      <th>Found by x tools</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6945</td>\n",
       "      <td>Q9H6L2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6945</td>\n",
       "      <td>Q8N5C7</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6945</td>\n",
       "      <td>Q9BS26</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7070</td>\n",
       "      <td>P57081</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6945</td>\n",
       "      <td>Q5T280</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>6945</td>\n",
       "      <td>O94817</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>45351</td>\n",
       "      <td>P12004</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>6945</td>\n",
       "      <td>O14653</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>45351</td>\n",
       "      <td>Q9NVU0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>45351</td>\n",
       "      <td>Q9NQ50</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Species UniProtID  Found by x tools\n",
       "0     6945    Q9H6L2                 1\n",
       "1     6945    Q8N5C7                 4\n",
       "2     6945    Q9BS26                 3\n",
       "3     7070    P57081                 4\n",
       "4     6945    Q5T280                 4\n",
       "..     ...       ...               ...\n",
       "72    6945    O94817                 3\n",
       "73   45351    P12004                 4\n",
       "74    6945    O14653                 4\n",
       "75   45351    Q9NVU0                 2\n",
       "76   45351    Q9NQ50                 3\n",
       "\n",
       "[77 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de9a79be-694f-417b-bd07-75962851d2fe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
