{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "065f3410-d7ed-454c-9b24-7e8e762d6e6b",
   "metadata": {},
   "source": [
    "# Mapping by genomic location"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31901541-5b08-4b78-bc20-76164ae7458f",
   "metadata": {},
   "source": [
    "1) open pkl files and save as dict\n",
    "2) open gff file\n",
    "3) parse gff file, save gene ids and check if CDS positions overlap\n",
    "4) compute overlap and save gene ids and overlap size for all\n",
    "5) filter overlaps by overlap size -> which overlap length shold we use?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ed16e3b9-121a-4388-9050-073e513b2e01",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import re\n",
    "import pandas as pd\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a989b420-086a-4186-9f97-45674251ef2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_pkl(path):\n",
    "    \n",
    "    pkl_file = open(path, 'rb')\n",
    "    dictionary = pickle.load(pkl_file)\n",
    "    pkl_file.close()\n",
    "    \n",
    "    return dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "59711d68-cf15-4fde-ba22-d7a6245c7a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "busco_augustus_pkl = open_pkl('../pkl_files/busco_augustus_CDS_positions_galga_v2.pkl')\n",
    "busco_metaeuk_pkl = open_pkl('../pkl_files/busco_metaeuk_CDS_positions_galga_v2.pkl')\n",
    "busco_augustus_species_pkl = open_pkl('../pkl_files/busco_augustus_with_species_parameter_CDS_positions_galga_v2.pkl')\n",
    "fdog_augustus_pkl = open_pkl('../pkl_files/fdog_assembly_augustus_busco_CDS_positions_galga_v2.pkl')\n",
    "fdog_metaeuk_pkl = open_pkl('../pkl_files/fdog_assembly_metaeuk_busco_CDS_positions_galga_v2.pkl')\n",
    "fdog_metaeuk_sens_pkl = open_pkl('../pkl_files/fdog_assembly_metaeuk_busco_CDS_positions_galga_v2_sens.pkl')\n",
    "compleasm_pkl = open_pkl('../pkl_files/compleasm_CDS_positions_galga_v2.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4f37f05-7dfc-4b34-8abe-0c7936389738",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Example NEMVE from ENA -> mapping with gene att: Name= or CDS att: protein_id=\n",
    "DS480309.1      Genbank region  1       1000    .       +       .       ID=DS480309.1:1..1000;Dbxref=taxon:45351;dev-stage=larval;gbkey=Src;genome=genomic;mol_type=genomic DNA;strain=CH2 x CH6\n",
    "DS480309.1      Genbank gene    668     847     .       +       .       ID=gene-NEMVEDRAFT_v1g157624;Name=NEMVEDRAFT_v1g157624;end_range=847,.;gbkey=Gene;gene_biotype=protein_coding;locus_tag=NEMVEDRAFT_v1g157624;partial=true;start_range=.,668\n",
    "DS480309.1      Genbank mRNA    668     847     .       +       .       ID=rna-NEMVEDRAFT_v1g157624;Parent=gene-NEMVEDRAFT_v1g157624;end_range=847,.;gbkey=mRNA;locus_tag=NEMVEDRAFT_v1g157624;partial=true;product=predicted protein;start_range=.,668\n",
    "DS480309.1      Genbank exon    668     847     .       +       .       ID=exon-NEMVEDRAFT_v1g157624-1;Parent=rna-NEMVEDRAFT_v1g157624;end_range=847,.;gbkey=mRNA;locus_tag=NEMVEDRAFT_v1g157624;partial=true;product=predicted protein;start_range=.,668\n",
    "DS480309.1      Genbank CDS     668     847     .       +       0       ID=cds-EDO25302.1;Parent=rna-NEMVEDRAFT_v1g157624;Dbxref=NCBI_GP:EDO25302.1;Name=EDO25302.1;gbkey=CDS;locus_tag=NEMVEDRAFT_v1g157624;partial=true;product=predicted protein;protein_id=EDO25302.1;start_range=.,668"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "602f2173-6dea-4cbf-9659-a5428c072750",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Example Ensembl 10116   Rattus norvegicus       UP000002494     Ensembl GCA_015227675.2\n",
    "#maping possible by gene att: ID=gene:\n",
    "1       mRatBN7.2       region  1       260522016       .       .       .       ID=region:1;Alias=CM026974.1,NC_051336.1\n",
    "1       ensembl gene    76834   358271  .       +       .       ID=gene:ENSRNOG00000070568;Name=Vom2r3;biotype=protein_coding;description=vomeronasal 2 receptor%2C 3 [Source:RGD Symbol%3BAcc:1565892];gene_id=ENSRNOG00000070568;version=1\n",
    "1       ensembl mRNA    76834   358271  .       +       .       ID=transcript:ENSRNOT00000096177;Parent=gene:ENSRNOG00000070568;Name=Vom2r3-201;biotype=protein_coding;tag=Ensembl_canonical;transcript_id=ENSRNOT00000096177;version=1\n",
    "1       ensembl five_prime_UTR  76834   76908   .       +       .       Parent=transcript:ENSRNOT00000096177\n",
    "1       ensembl exon    76834   77114   .       +       .       Parent=transcript:ENSRNOT00000096177;constitutive=1;exon_id=ENSRNOE00000506230;rank=1;version=2\n",
    "1       ensembl CDS     76909   77114   .       +       0       ID=CDS:ENSRNOP00000084716;Parent=transcript:ENSRNOT00000096177;protein_id=ENSRNOP00000084716;version=1\n",
    "1       ensembl exon    79753   80035   .       +       .       Parent=transcript:ENSRNOT00000096177;constitutive=1;exon_id=ENSRNOE00000620383;rank=2;version=1\n",
    "1       ensembl CDS     79753   80035   .       +       1       ID=CDS:ENSRNOP00000084716;Parent=transcript:ENSRNOT00000096177;protein_id=ENSRNOP00000084716;version=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "53e31769-1b74-45f8-a471-125fad1e1bdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gff_ena(path, sp, mapping_uniprot_id):\n",
    "    gff_file = open(path, 'r')\n",
    "    lines = gff_file.readlines()\n",
    "    gff_dic = {}\n",
    "    for line in lines:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        else:\n",
    "            contig, source, typ, start, end, score, strand, phase, att = line.split('\\t')\n",
    "            if typ == 'region':\n",
    "                gff_dic[contig] = {}\n",
    "            elif typ == 'gene':\n",
    "                gene_name = re.search(r'Name=(.*?);', att).group(1)\n",
    "                transcript = 0\n",
    "            elif typ == 'mRNA':\n",
    "                transcript += 1\n",
    "            elif typ == 'CDS':\n",
    "                # all CDs have the same ids so we can not save them as a dict\n",
    "                cds_names = set()\n",
    "                try:\n",
    "                    cds_name = re.search(r'Name=(.*?);', att).group(1)\n",
    "                    \n",
    "                except AttributeError:\n",
    "                    cds_name = re.search(r'ID=(.*?);', att).group(1)\n",
    "                cds_names.add(cds_name)\n",
    "                try:\n",
    "                    cds_locus = re.search(r'locus_tag=(.*?);', att).group(1)\n",
    "                except AttributeError:\n",
    "                    cds_locus = re.search(r'gene=(.*?);', att).group(1)\n",
    "                cds_names.add(cds_locus)\n",
    "                try:\n",
    "                    cds_standard = re.search(r'standard_name=(.*?$)', att).group(1)\n",
    "                    cds_names.add(cds_standard)\n",
    "                except AttributeError:\n",
    "                    pass\n",
    "                \n",
    "                try:\n",
    "                    gff_dic[contig][gene_name].append([int(start), int(end), score, strand, phase, cds_names, transcript])\n",
    "                except KeyError:\n",
    "                    gff_dic[contig][gene_name] = []\n",
    "                    transcript = 1\n",
    "                    gff_dic[contig][gene_name].append([int(start), int(end), score, strand, phase, cds_names, transcript])\n",
    "    return gff_dic           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "affef07a-de3d-4652-9e02-f5cfdc98f64d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gff_ensembl(path, sp, mapping_uniprot_id):\n",
    "    gff_file = open(path, 'r')\n",
    "    lines = gff_file.readlines()\n",
    "    gff_dic = {}\n",
    "    contig_mapping = {}\n",
    "    for line in lines:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        else:\n",
    "            contig, source, typ, start, end, score, strand, phase, att = line.split('\\t')\n",
    "            if typ == 'region':\n",
    "                #print(line)\n",
    "                try:\n",
    "                    alias = re.search(r';Alias=(.*?$)', att).group(1)\n",
    "                    for i in alias.split(','):\n",
    "                        contig_mapping[i] = contig\n",
    "                except AttributeError:\n",
    "                    pass\n",
    "                gff_dic[contig] = {}\n",
    "            elif typ == 'gene':\n",
    "                cds_names = set()\n",
    "                gene_name = re.search(r'ID=gene:(.*?);', att).group(1)\n",
    "                try:\n",
    "                    gene_alias = re.search(r';Name=(.*?);', att).group(1)\n",
    "                    cds_names.add(gene_alias)\n",
    "                except AttributeError:\n",
    "                    pass\n",
    "                try:\n",
    "                    entrenz_alias = re.search(r'BAcc:(.*?)]', att).group(1)\n",
    "                    cds_names.add(entrenz_alias)\n",
    "                except AttributeError:\n",
    "                    pass\n",
    "                cds_names.add(gene_name)\n",
    "                try:\n",
    "                    print(gff_dic[contig][gene_name])\n",
    "                    pass\n",
    "                except KeyError:\n",
    "                    try:\n",
    "                        gff_dic[contig][gene_name] = []\n",
    "                    except KeyError:\n",
    "                        try:\n",
    "                            gff_dic[contig][gene_name] = []\n",
    "                        except KeyError:\n",
    "                            gff_dic[contig] = {}\n",
    "                            gff_dic[contig][gene_name] = []\n",
    "            elif typ == 'ncRNA_gene':\n",
    "                cds_names = set()\n",
    "                gene_name = re.search(r'ID=gene:(.*?);', att).group(1)\n",
    "                try:\n",
    "                    gene_alias = re.search(r';Name=(.*?);', att).group(1)\n",
    "                    cds_names.add(gene_alias)\n",
    "                except AttributeError:\n",
    "                    pass\n",
    "                cds_names.add(gene_name)\n",
    "                #print(line)\n",
    "                try:\n",
    "                    print(gff_dic[contig][gene_name])\n",
    "                except KeyError:\n",
    "                    try:\n",
    "                        gff_dic[contig][gene_name] = []\n",
    "                    except KeyError:\n",
    "                        try:\n",
    "                            gff_dic[contig][gene_name] = []\n",
    "                        except KeyError:\n",
    "                            gff_dic[contig] = {}\n",
    "            elif typ == 'mRNA':\n",
    "                transcript = re.search(r'ID=transcript:(.*?);', att).group(1)\n",
    "            elif typ == 'CDS':\n",
    "                # all CDs have the same ids so we can not save them as a dict\n",
    "                cds_name = re.search(r'ID=CDS:(.*?);', att).group(1)\n",
    "                cds_names.add(cds_name)\n",
    "                try:\n",
    "                    gff_dic[contig][gene_name].append([int(start), int(end), score, strand, phase, cds_names, transcript])\n",
    "                except KeyError:\n",
    "                    print(line)\n",
    "                    gff_dic[contig][gene_name].append([int(start), int(end), score, strand, phase, cds_names, transcript])\n",
    "    return gff_dic, contig_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7eeec3c1-8e25-406b-b9ed-9a5494e840ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_species_file(file):\n",
    "    lines = file.readlines()\n",
    "    species_dict = {}\n",
    "    for line in lines:\n",
    "        line = line.rstrip()\n",
    "        ncbi, name, uniprot_acc, source, refseq_acc = line.split('\\t')\n",
    "        species_dict[ncbi] = {'name': name, 'uniprot': uniprot_acc, 'source': source, 'refseq': refseq_acc}\n",
    "    return species_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a78076bf-812f-4d88-8704-89fa81fb56c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_uniprot(path):\n",
    "    uniprot_mapping_dict = {}\n",
    "    file = open(path, 'r')\n",
    "    lines = file.readlines()\n",
    "    counter = 0\n",
    "    #not_unique_source = {'ZFIN','Xenbase','RGD','CRC64','GI','UniParc','OMA','ESTHER','MEROPS','Gene_ORFName','DNASU','Gene_Synonym', 'UniRef100', 'UniRef90', 'UniRef50', 'NCBI_TaxID', 'OrthoDB', 'EMBL', 'EMBL-CDS', 'VEuPathDB', 'eggNOG', 'GeneTree', 'ChiTaRS', 'HOGENOM', 'UniPathway','GlyConnect', 'GuidetoPHARMACOLOGY', 'DrugBank', 'BioGRID', 'UCSC', 'PDB', 'ComplexPortal', 'TCDB', 'ChEMBL', 'DIP', 'KEGG', 'TreeFam', 'Reactome', 'STRING', 'Allergome'}\n",
    "    for line in lines:\n",
    "        line = line.rstrip()\n",
    "        uniprot, source, source_id = line.split('\\t')\n",
    "        #if source in not_unique_source:\n",
    "         #   continue\n",
    "        if source.startswith('Ensembl'):\n",
    "            # some of the ensample names also have versions given at the end of the id seperated by a dot. But the version number is often not given in the gff gene name\n",
    "            prefix_source_id = source_id.split('.')[0]\n",
    "            try:\n",
    "                uniprot_mapping_dict[prefix_source_id].add(uniprot)\n",
    "            except KeyError:\n",
    "                uniprot_mapping_dict[prefix_source_id] = set()\n",
    "                uniprot_mapping_dict[prefix_source_id].add(uniprot)\n",
    "        try:\n",
    "            uniprot_mapping_dict[source_id].add(uniprot)\n",
    "        except KeyError:\n",
    "            uniprot_mapping_dict[source_id] = set()\n",
    "            uniprot_mapping_dict[source_id].add(uniprot)\n",
    "    return uniprot_mapping_dict\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f935d20d-8ec3-45bd-9a4f-9a1491462ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in assembly information\n",
    "#species_file = open('../../data/fDOG-assembly/species_set_benchmark.tsv', 'r')\n",
    "species_file = open('../../data/fDOG-assembly/species_set_benchmark_v2.tsv', 'r')\n",
    "species_dict = parse_species_file(species_file)\n",
    "species_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "87eae7ff-3bab-4bf7-9176-c571c7dc09c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45351\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_000209225.1/genomic.gff\n",
      "ENA\n",
      "10116\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_015227675.2/genes.gff3\n",
      "Ensembl\n",
      "9031\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_000002315.5/genes.gff3\n",
      "Ensembl\n",
      "8364\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCF_000004195.4/genomic.gff\n",
      "ENA\n",
      "7955\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCF_000002035.6/genomic.gff\n",
      "ENA\n",
      "7227\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_000001215.4/genomic.gff\n",
      "ENA\n",
      "7070\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_000002335.3/genomic.gff\n",
      "ENA\n",
      "6945\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_000208615.1/genes.gff3\n",
      "Ensembl\n",
      "ENA\n",
      "6412\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_000326865.1/genes.gff3\n",
      "Ensembl\n",
      "ENA\n",
      "6239\n",
      "../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/GCA_000002985.3/genomic.gff\n",
      "ENA\n"
     ]
    }
   ],
   "source": [
    "gff_path  = '../../data/qfo_eukaryota_2022/ncbi_download/ncbi_dataset/data/'\n",
    "positions_gff = {}\n",
    "contig_mapping_dict = {}\n",
    "mapping_files_path = '/share/gluster/Projects/hannah/fDOG-assembly/benchmark/data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/'\n",
    "for sp in species_dict:\n",
    "    print(sp)\n",
    "    mapping_uniprot_ids = mapping_uniprot(mapping_files_path + species_dict[sp]['uniprot'] + '_' + sp + '.idmapping')\n",
    "    if species_dict[sp]['source'] == 'Ensembl':\n",
    "        print(gff_path + species_dict[sp]['refseq']+ '/genes.gff3')\n",
    "        try:\n",
    "            print('Ensembl')\n",
    "            species_gff_dict, contig_mapping = parse_gff_ensembl(gff_path + species_dict[sp]['refseq']+ '/genes.gff3', sp, mapping_uniprot_ids)\n",
    "        except FileNotFoundError:\n",
    "            print('ENA')\n",
    "            species_gff_dict = parse_gff_ena(gff_path + species_dict[sp]['refseq']+ '/genomic.gff', sp, mapping_uniprot_ids)\n",
    "        positions_gff[sp] = species_gff_dict\n",
    "        contig_mapping_dict[sp] = contig_mapping\n",
    "    else:\n",
    "        print(gff_path + species_dict[sp]['refseq']+ '/genomic.gff')\n",
    "        print('ENA')\n",
    "        species_gff_dict = parse_gff_ena(gff_path + species_dict[sp]['refseq']+ '/genomic.gff', sp, mapping_uniprot_ids)\n",
    "        positions_gff[sp] = species_gff_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "54a750c9-bf3a-4edc-a0f6-0a098907cc19",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_ref_positional_dict(contig_gff):\n",
    "    # input dict structure: positions_gff[ncbi][contig][gene][transcripts_list[start,end,score, strand, phase, cds_names, transcript]] \n",
    "    # should sort genes according to start position on contig\n",
    "    sorted_dict = dict(sorted(contig_gff.items(), key=lambda item: item[0][0]))\n",
    "    return sorted_dict\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7afde679-12e1-416d-9c1a-9b03a0c393cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for species in positions_gff:\n",
    "    for contig in positions_gff[species]:\n",
    "        positions_gff[species][contig] = sort_ref_positional_dict(positions_gff[species][contig])\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7b6d4672-f324-47d2-85e2-c10ca0439409",
   "metadata": {},
   "outputs": [],
   "source": [
    "def give_overlap(start, end, strand, start_r, end_r, strand_r):\n",
    "    overlap = 0\n",
    "    if strand != strand_r:\n",
    "        return overlap\n",
    "    elif end > start_r and end <= end_r:\n",
    "        overlap = end - max(start_r, start)\n",
    "    elif start >= start_r and start < end_r:\n",
    "        overlap = min(end_r, end) - start\n",
    "    return overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "72802b6a-e31c-463f-9834-388f8082d1d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_coding_length(cds_list):\n",
    "    cds_dict = {}\n",
    "    for cds_ref in cds_list:\n",
    "        start_r, end_r, score_r, strand_r, phase_r, cds_name_r, transcript_id = cds_ref\n",
    "        try:\n",
    "            cds_dict[transcript_id] += int(end_r) - int(start_r)\n",
    "        except KeyError:\n",
    "            cds_dict[transcript_id] = int(end_r) - int(start_r)\n",
    "    return cds_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "037f49b8-eedb-45f4-a02a-f232ebd93cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlap_table(pkl_file, positions_gff = positions_gff):    \n",
    "    mapping_table = []\n",
    "    mapping_files_path = '/share/gluster/Projects/hannah/fDOG-assembly/benchmark/data/qfo_eukaryota_2022/qfo_data_2022/Eukaryota/'\n",
    "    overlap_dict = {'Species': [], 'GeneID':[], 'transcript':[], 'overlap': [], 'uniprotID': [], 'coverage': []}\n",
    "    cant_map = set()\n",
    "    multi_map = set()\n",
    "    check = False\n",
    "    p = False\n",
    "    statistic_dict = {'Species': [], '#Orthologs': [], 'No_overlap': []}\n",
    "    for key in pkl_file:\n",
    "        # no overlap with any CDS of a gene in ref annotation file\n",
    "        no_overlap = 0\n",
    "        # number of orthologs predicted in total\n",
    "        gene_count = 0\n",
    "        cant_map = set()\n",
    "        multi_map = set()\n",
    "        print(key)\n",
    "        ncbi = key.split('@')[1]\n",
    "        #if key == 'GALGA@9031@000002315_5':\n",
    "        if key == 'GALGA@9031@QfO22':\n",
    "           continue\n",
    "        ######## for human proteom bechmark ###########\n",
    "        #if key != 'RATNO@10116@QfO22' and key != 'NEMVE@45351@QfO22' :\n",
    "         #   print('Not right key')\n",
    "          #  print(key)\n",
    "           # continue\n",
    "        ################testing###########\n",
    "        #if key != 'RATNO@10116@QfO22':\n",
    "         #   continue\n",
    "        ########### teating############\n",
    "        #dict scheme: position_dict_cds[ass]:[contig]:[geneid]:[transcript_id]:list([source, typ, int(start), int(end), strand, phase, att])\n",
    "        #dict scheme contig_mapping_dict: gff_dic[species]:[contig]:[gene_name]:list([start, end, score, strand, phase, cds_name])\n",
    "        print(species_dict[ncbi])\n",
    "        mapping_uniprot_ids = mapping_uniprot(mapping_files_path + species_dict[ncbi]['uniprot'] + '_' + ncbi + '.idmapping')\n",
    "        for contig in pkl_file[key]:\n",
    "            try:\n",
    "                genes = positions_gff[ncbi][contig]\n",
    "                contig_ref = contig\n",
    "            except KeyError:\n",
    "                try:\n",
    "                    alias = contig_mapping_dict[ncbi][contig]\n",
    "                    genes = positions_gff[ncbi][alias]\n",
    "                    contig_ref = alias\n",
    "                except KeyError:\n",
    "                    print(key)\n",
    "                    print(contig)\n",
    "                    print(ncbi)\n",
    "                    print(species_dict[ncbi])\n",
    "            for gene in pkl_file[key][contig]:\n",
    "                ###### testing #############\n",
    "                #if gene == '147873at33208':\n",
    "                 #   p = True\n",
    "                  #  print('Gene found in pkl')\n",
    "                #else:\n",
    "                 #   p = False\n",
    "                ###### testing #############\n",
    "                for transcript in pkl_file[key][contig][gene]:\n",
    "                    cds_list = pkl_file[key][contig][gene][transcript]\n",
    "                    ###### testing #########\n",
    "                    #if p == True:\n",
    "                    #    print('CDS list')\n",
    "                    #    print(cds_list)\n",
    "                    #else:\n",
    "                    #    continue\n",
    "                    ###### testing #########\n",
    "                    overlaps = {'uniprotID':[], 'overlap':[], 'transcript_id': [], 'gene_ref_name':[]}\n",
    "                    length_transcript = 0\n",
    "                    #cds_sum_length_dict = {}\n",
    "                    cds_sum_dict = {}\n",
    "                    for cds in cds_list:\n",
    "                        #try:\n",
    "                        source, typ, start, end, strand, phase, att = cds\n",
    "                        length_transcript += end - start\n",
    "                        #except ValueError:\n",
    "                         #   print(cds)\n",
    "                        for gene_ref in positions_gff[ncbi][contig_ref]:\n",
    "                            #if check == True:\n",
    "                              #  check = False\n",
    "                              #  break\n",
    "                            for cds_ref in positions_gff[ncbi][contig_ref][gene_ref]:\n",
    "                                try:\n",
    "                                    start_r, end_r, score_r, strand_r, phase_r, cds_name_r, transcript_id = cds_ref\n",
    "                                except ValueError:\n",
    "                                    print(cds_ref)\n",
    "                                try:\n",
    "                                    cds_sum_dict[transcript_id]\n",
    "                                except KeyError:\n",
    "                                    cds_sum_dict[transcript_id] = int(end_r) - int(start_r)\n",
    "                                #if min(int(end_r),int(start_r)) > max(int(end),int(start)):\n",
    "                                    #check = True\n",
    "                                    #break\n",
    "                                    #is not working because the dictionaries are not correctly sorted\n",
    "                                overlap = give_overlap(int(start), int(end), strand, int(start_r), int(end_r), strand_r)\n",
    "                                if overlap != 0:\n",
    "                                    multi = False\n",
    "                                    if type(cds_name_r) == set:\n",
    "                                        uniprot_id_r = None\n",
    "                                        for i in cds_name_r:\n",
    "                                            try:\n",
    "                                                uniprot_ids = mapping_uniprot_ids[i]\n",
    "                                                if len(uniprot_ids) > 1:\n",
    "                                                    multi = True\n",
    "                                                else:\n",
    "                                                    uniprot_id_r = list(uniprot_ids)[0]\n",
    "                                                    break\n",
    "                                            except KeyError:\n",
    "                                                pass\n",
    "                                        if uniprot_id_r != None:\n",
    "                                            overlaps['uniprotID'].append(uniprot_id_r)\n",
    "                                        elif uniprot_id_r == None and multi == True:\n",
    "                                            uniprot_id_r = 'NA'\n",
    "                                            multi_map.add(next(iter(cds_name_r)))\n",
    "                                            overlaps['uniprotID'].append(uniprot_id_r)\n",
    "                                        else:\n",
    "                                            #print(cds_name_r)\n",
    "                                            if uniprot_id_r == None:\n",
    "                                                uniprot_id_r = 'NA'\n",
    "                                            overlaps['uniprotID'].append(uniprot_id_r)\n",
    "                                            cant_map.add(next(iter(cds_name_r)))\n",
    "                                    else:\n",
    "                                        try:\n",
    "                                            uniprot_ids = mapping_uniprot_ids[i]\n",
    "                                            if len(uniprot_ids) > 1:\n",
    "                                                    multi = True\n",
    "                                                    multi_map.add(next(iter(cds_name_r)))\n",
    "                                                    overlaps['uniprotID'].append('NA')\n",
    "                                            else:\n",
    "                                                uniprot_id_r = list(uniprot_ids)[0]\n",
    "                                                overlaps['uniprotID'].append(uniprot_id_r)\n",
    "                                        except KeyError: \n",
    "                                            cant_map.add(cds_name_r)\n",
    "                                            overlaps['uniprotID'].append('NA')\n",
    "                                    overlaps['overlap'].append(overlap)\n",
    "                                    overlaps['transcript_id'].append(transcript_id)\n",
    "                                    overlaps['gene_ref_name'].append(gene_ref)\n",
    "                            cds_sum_dict[gene_ref] = get_coding_length(positions_gff[ncbi][contig_ref][gene_ref])\n",
    "                    #print(overlaps)\n",
    "                    df = pd.DataFrame(overlaps)\n",
    "                    ###### testing #############\n",
    "                    #if p == True:\n",
    "                    #    print('overlaps')\n",
    "                    #    print(df)\n",
    "                    ###### testing #############\n",
    "                    df_sum = df.groupby(['uniprotID', 'transcript_id', 'gene_ref_name']).sum()\n",
    "                    ##### Only Max Uniprot###################\n",
    "                    #max_ids = df_sum.reset_index().iloc[df_sum.reset_index().overlap.idxmax()]\n",
    "                    #print(max_ids)\n",
    "                    #overlap_dict['Species'].append(ncbi)\n",
    "                    #overlap_dict['GeneID'].append(gene)\n",
    "                    #overlap_dict['TranscriptID'].append(transcript)\n",
    "                    #overlap_dict['overlap'].append(max_ids['overlap'])\n",
    "                    #overlap_dict['uniprotID'].append(max_ids['uniprotID'])\n",
    "                    #print(overlap_dict)\n",
    "                    #### Max per Uniprot#######################\n",
    "                    max_length = df_sum.groupby(['uniprotID', 'transcript_id', 'gene_ref_name'])['overlap'].max()\n",
    "                    sum_lengths_list = max_length.reset_index().values.tolist()\n",
    "                    gene_count += 1\n",
    "                    ###### testing #############\n",
    "                    #if p == True:\n",
    "                    #    print('df_sum')\n",
    "                    #    print(df_sum)\n",
    "                    #    print('max_length')\n",
    "                    #    print(max_length)\n",
    "                    #    print('sum_lenght_list')\n",
    "                    #    print(sum_lengths_list)\n",
    "                    ###### testing #############\n",
    "                    for row in sum_lengths_list:\n",
    "                        #print(row)\n",
    "                        overlap_dict['Species'].append(ncbi)\n",
    "                        overlap_dict['GeneID'].append(gene)\n",
    "                        overlap_dict['transcript'].append(transcript)\n",
    "                        overlap_dict['overlap'].append(row[3])\n",
    "                        if row[0] == 'NA':\n",
    "                            overlap_dict['uniprotID'].append(None)\n",
    "                        else: \n",
    "                            overlap_dict['uniprotID'].append(row[0])\n",
    "                        overlap_dict['coverage'].append(int(row[3])/min(length_transcript,cds_sum_dict[row[2]][row[1]]))\n",
    "                        cov = int(row[3])/min(length_transcript,cds_sum_dict[row[2]][row[1]])\n",
    "                        if cov > 1:\n",
    "                            print(ncbi)\n",
    "                            print(gene)\n",
    "                            print(row)\n",
    "                            print(length_transcript)\n",
    "                            print(cds_sum_dict[row[2]][row[1]])\n",
    "                            print(cds_list)\n",
    "                            print(df)\n",
    "                            print(cov)\n",
    "                    if overlaps['uniprotID'] == []:\n",
    "                        #print(ncbi)\n",
    "                        #print(gene)\n",
    "                        no_overlap += 1\n",
    "                        overlap_dict['Species'].append(ncbi)\n",
    "                        overlap_dict['GeneID'].append(gene)\n",
    "                        overlap_dict['transcript'].append(transcript)\n",
    "                        overlap_dict['overlap'].append(None)\n",
    "                        overlap_dict['uniprotID'].append(None)\n",
    "                        overlap_dict['coverage'].append(None)\n",
    "        \n",
    "        print(f'Gene_count %d, No overlap %d, Cant map set %d, Multi map set %d'% (gene_count, no_overlap, len(cant_map), len(multi_map)))\n",
    "        print('Multi map set')\n",
    "        print(multi_map)\n",
    "        print('Cant map examples')\n",
    "        if len(cant_map) > 3:\n",
    "            print(random.sample(sorted(cant_map), 3))\n",
    "        else:\n",
    "            print(cant_map)\n",
    "        #statistic_dict = {'Species': [], '#Orthologs': [], 'No_overlap': []}\n",
    "        statistic_dict['Species'].append(ncbi)\n",
    "        statistic_dict['#Orthologs'].append(gene_count)\n",
    "        statistic_dict['No_overlap'].append(no_overlap)\n",
    "    df_stat = pd.DataFrame(statistic_dict)\n",
    "    display(df_stat)\n",
    "        \n",
    "    #print(len(overlap_dict['Species']), len(overlap_dict['GeneID']), len(overlap_dict['transcript']), len(overlap_dict['overlap']), len(overlap_dict['uniprotID']), len(overlap_dict[ 'coverage']))\n",
    "    df_overlap = pd.DataFrame(overlap_dict)\n",
    "    \n",
    "    return df_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b205b093-3a17-4c0f-a9ca-5f98aeb7cc71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compleasm\n",
      "CAEEL@6239@QfO22\n",
      "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n",
      "Gene_count 729, No overlap 0, Cant map set 0, Multi map set 3\n",
      "Multi map set\n",
      "{'CELE_Y39A3CR.1', 'Y39A3CR.1d', 'Y39A3CR.1b'}\n",
      "Cant map examples\n",
      "set()\n",
      "DANRE@7955@QfO22\n",
      "{'name': 'Danio rerio', 'uniprot': 'UP000000437', 'source': 'RefSeq', 'refseq': 'GCF_000002035.6'}\n",
      "Gene_count 1155, No overlap 1, Cant map set 0, Multi map set 1\n",
      "Multi map set\n",
      "{'NP_001017751.2'}\n",
      "Cant map examples\n",
      "set()\n",
      "DROME@7227@QfO22\n",
      "{'name': 'Drosophila melanogaster', 'uniprot': 'UP000000803', 'source': 'ENA/EMBL', 'refseq': 'GCA_000001215.4'}\n",
      "Gene_count 957, No overlap 1, Cant map set 7, Multi map set 1\n",
      "Multi map set\n",
      "{'Dmel_CG7843'}\n",
      "Cant map examples\n",
      "['Dmel_CG9742', 'Dmel_CG1821', 'AAF48554.1']\n",
      "GALGA@9031@000002315_5\n",
      "{'name': 'Gallus gallus', 'uniprot': 'UP000000539', 'source': 'Ensembl', 'refseq': 'GCA_000002315.5'}\n",
      "Gene_count 918, No overlap 8, Cant map set 18, Multi map set 2\n",
      "Multi map set\n",
      "{'MRPL40', 'ENSGALP00000090837'}\n",
      "Cant map examples\n",
      "['ENSGALG00000059518', 'ENSGALG00000056104', 'ENSGALP00000093534']\n",
      "HELRO@6412@QfO22\n",
      "{'name': 'Helobdella robusta', 'uniprot': 'UP000015101', 'source': 'Ensembl', 'refseq': 'GCA_000326865.1'}\n",
      "Gene_count 921, No overlap 22, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "IXOSC@6945@QfO22\n",
      "{'name': 'Ixodes scapularis', 'uniprot': 'UP000001555', 'source': 'Ensembl', 'refseq': 'GCA_000208615.1'}\n",
      "Gene_count 928, No overlap 67, Cant map set 17, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['EEC05118.1', 'EEC06299.1', 'EEC11981.1']\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 948, No overlap 28, Cant map set 37, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['cds-NEMVEDRAFT_v1g113129', 'NEMVEDRAFT_v1g21521', 'NEMVEDRAFT_v1g52726']\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 1060, No overlap 33, Cant map set 1, Multi map set 10\n",
      "Multi map set\n",
      "{'ENSRNOP00000006063', 'ENSRNOG00000012287', 'ENSRNOP00000075579', 'ENSRNOG00000018680', 'ENSRNOP00000061132', '620870', 'ENSRNOP00000058286', 'ENSRNOG00000063018', '1309282', 'ENSRNOG00000010587'}\n",
      "Cant map examples\n",
      "{'9247033'}\n",
      "TRICA@7070@QfO22\n",
      "{'name': 'Tribolium castaneum', 'uniprot': 'UP000007266', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002335.3'}\n",
      "Gene_count 955, No overlap 3, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "XENTR@8364@QfO22\n",
      "{'name': 'Xenopus tropicalis', 'uniprot': 'UP000008143', 'source': 'RefSeq', 'refseq': 'GCF_000004195.4'}\n",
      "Gene_count 958, No overlap 1, Cant map set 0, Multi map set 49\n",
      "Multi map set\n",
      "{'XP_012814324.1', 'NP_001007514.1', 'mrpl47', 'NP_989259.2', 'XP_002938554.1', 'XP_012822075.1', 'NP_001016494.1', 'ctu2', 'XP_002931922.1', 'NP_001027490.1', 'XP_012822063.1', 'rrn3', 'poldip2', 'supt4h1', 'XP_012814321.1', 'lsm7', 'XP_004912704.1', 'XP_004913000.1', 'XP_004912995.1', 'mrps5', 'XP_017950881.1', 'elp4', 'XP_012827224.2', 'pwp1', 'XP_004912997.1', 'XP_004912577.1', 'XP_004912998.1', 'XP_002933349.1', 'c12orf4', 'NP_001016895.1', 'fdxr', 'XP_012814325.1', 'XP_017947959.1', 'XP_004912705.1', 'NP_001016714.1', 'XP_017951362.1', 'ift52', 'tm2d3', 'tbc1d7', 'vps54', 'bbs4', 'XP_002937119.1', 'XP_004912996.1', 'XP_012809948.1', 'XP_017947958.1', 'utp18', 'c1qbp', 'tmem209', 'XP_002934120.1'}\n",
      "Cant map examples\n",
      "set()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6239</td>\n",
       "      <td>729</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7955</td>\n",
       "      <td>1155</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7227</td>\n",
       "      <td>957</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9031</td>\n",
       "      <td>918</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6412</td>\n",
       "      <td>921</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6945</td>\n",
       "      <td>928</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>45351</td>\n",
       "      <td>948</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10116</td>\n",
       "      <td>1060</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7070</td>\n",
       "      <td>955</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8364</td>\n",
       "      <td>958</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0    6239         729           0\n",
       "1    7955        1155           1\n",
       "2    7227         957           1\n",
       "3    9031         918           8\n",
       "4    6412         921          22\n",
       "5    6945         928          67\n",
       "6   45351         948          28\n",
       "7   10116        1060          33\n",
       "8    7070         955           3\n",
       "9    8364         958           1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(\"Compleasm\")\n",
    "compleasm_df = get_overlap_table(compleasm_pkl)\n",
    "compleasm_df.to_csv('../overlap_tables/compleasm_overlap_gff_files_gallus_v2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8b81729f-fb69-429d-a9e3-716d38e70682",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BSUCO augustus species\n",
      "CAEEL@6239@QfO22\n",
      "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n",
      "Gene_count 731, No overlap 1, Cant map set 10, Multi map set 3\n",
      "Multi map set\n",
      "{'Y39A3CR.1c', 'Y39A3CR.1d', 'Y39A3CR.1b'}\n",
      "Cant map examples\n",
      "['cds-CELE_Y71A12C.3', 'Y57G11C.21', 'cds-CELE_F27D4.3']\n",
      "DANRE@7955@QfO22\n",
      "{'name': 'Danio rerio', 'uniprot': 'UP000000437', 'source': 'RefSeq', 'refseq': 'GCF_000002035.6'}\n",
      "Gene_count 1241, No overlap 16, Cant map set 0, Multi map set 1\n",
      "Multi map set\n",
      "{'NP_001017751.2'}\n",
      "Cant map examples\n",
      "set()\n",
      "DROME@7227@QfO22\n",
      "{'name': 'Drosophila melanogaster', 'uniprot': 'UP000000803', 'source': 'ENA/EMBL', 'refseq': 'GCA_000001215.4'}\n",
      "Gene_count 958, No overlap 0, Cant map set 8, Multi map set 2\n",
      "Multi map set\n",
      "{'Dmel_CG7843', 'AAF49536.1'}\n",
      "Cant map examples\n",
      "['AAF58920.1', 'AAF48634.1', 'AAS65037.3']\n",
      "GALGA@9031@000002315_5\n",
      "{'name': 'Gallus gallus', 'uniprot': 'UP000000539', 'source': 'Ensembl', 'refseq': 'GCA_000002315.5'}\n",
      "Gene_count 912, No overlap 5, Cant map set 20, Multi map set 2\n",
      "Multi map set\n",
      "{'TRAPPC2L', 'ENSGALP00000074383'}\n",
      "Cant map examples\n",
      "['ENSGALG00000064458', 'ENSGALG00000056352', 'ENSGALP00000079331']\n",
      "HELRO@6412@QfO22\n",
      "{'name': 'Helobdella robusta', 'uniprot': 'UP000015101', 'source': 'Ensembl', 'refseq': 'GCA_000326865.1'}\n",
      "Gene_count 894, No overlap 16, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "IXOSC@6945@QfO22\n",
      "{'name': 'Ixodes scapularis', 'uniprot': 'UP000001555', 'source': 'Ensembl', 'refseq': 'GCA_000208615.1'}\n",
      "Gene_count 889, No overlap 67, Cant map set 14, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['IscW_ISCW004894', 'IscW_ISCW007765', 'EEC11981.1']\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 931, No overlap 36, Cant map set 33, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['cds-NEMVEDRAFT_v1g21521', 'NEMVEDRAFT_v1g115845', 'cds-NEMVEDRAFT_v1g10259']\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 1018, No overlap 60, Cant map set 1, Multi map set 8\n",
      "Multi map set\n",
      "{'Ak6', 'ENSRNOG00000001172', 'ENSRNOP00000061132', '621291', 'ENSRNOP00000006063', 'ENSRNOP00000016399', 'Hspbap1', 'ENSRNOP00000058286'}\n",
      "Cant map examples\n",
      "{'ENSRNOP00000049054'}\n",
      "TRICA@7070@QfO22\n",
      "{'name': 'Tribolium castaneum', 'uniprot': 'UP000007266', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002335.3'}\n",
      "Gene_count 956, No overlap 3, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "XENTR@8364@QfO22\n",
      "{'name': 'Xenopus tropicalis', 'uniprot': 'UP000008143', 'source': 'RefSeq', 'refseq': 'GCF_000004195.4'}\n",
      "Gene_count 958, No overlap 1, Cant map set 0, Multi map set 48\n",
      "Multi map set\n",
      "{'NP_001165133.1', 'c12orf4', 'gbf1', 'XP_004918010.1', 'tbc1d7', 'XP_002939365.2', 'tm2d3', 'NP_001007514.1', 'NP_989259.2', 'XP_012827224.2', 'rrn3', 'rpl28', 'NP_001016679.1', 'tmem209', 'XP_002934120.1', 'det1', 'utp18', 'XP_002931922.1', 'XP_004912996.1', 'crnkl1', 'XP_012822063.1', 'XP_004912998.1', 'XP_017947958.1', 'XP_004914765.1', 'ift52', 'NP_001017098.1', 'NP_001016494.1', 'ankmy2', 'NP_001016280.1', 'XP_004912997.1', 'XP_012814323.1', 'vps41', 'XP_017950881.1', 'NP_001011344.1', 'ctu2', 'XP_017947959.1', 'gtf2h1', 'XP_004912705.1', 'mrps5', 'mrpl47', 'c1qbp', 'NP_001027490.1', 'XP_002938755.1', 'NP_001121527.1', 'nup107', 'pwp1', 'mst1r', 'tmem70'}\n",
      "Cant map examples\n",
      "set()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6239</td>\n",
       "      <td>731</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7955</td>\n",
       "      <td>1241</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7227</td>\n",
       "      <td>958</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9031</td>\n",
       "      <td>912</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6412</td>\n",
       "      <td>894</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6945</td>\n",
       "      <td>889</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>45351</td>\n",
       "      <td>931</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10116</td>\n",
       "      <td>1018</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7070</td>\n",
       "      <td>956</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8364</td>\n",
       "      <td>958</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0    6239         731           1\n",
       "1    7955        1241          16\n",
       "2    7227         958           0\n",
       "3    9031         912           5\n",
       "4    6412         894          16\n",
       "5    6945         889          67\n",
       "6   45351         931          36\n",
       "7   10116        1018          60\n",
       "8    7070         956           3\n",
       "9    8364         958           1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(\"BSUCO augustus species\")\n",
    "busco_augustus_df = get_overlap_table(busco_augustus_species_pkl)\n",
    "busco_augustus_df.to_csv('../overlap_tables/busco_augustus_species_overlap_gff_files_gallus_v2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72cea962-9726-4241-a652-fc6c0d297d36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BUSCO augustus\n",
      "CAEEL@6239@QfO22\n",
      "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n"
     ]
    }
   ],
   "source": [
    "print(\"BUSCO augustus\")\n",
    "busco_augustus_df = get_overlap_table(busco_augustus_pkl)\n",
    "#busco_augustus_df.to_csv('../overlap_tables/busco_augustus_overlap_gff_files_gallus_v2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2384f1e4-8532-4a31-806f-2b493b0c014f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BUSCO metaeuk\n",
      "CAEEL@6239@QfO22\n",
      "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n",
      "Gene_count 739, No overlap 1, Cant map set 4, Multi map set 3\n",
      "Multi map set\n",
      "{'Y39A3CR.1c', 'CAH2176098.1', 'CAH2176100.1'}\n",
      "Cant map examples\n",
      "['cds-CELE_B0250.3', 'F59E12.16b', 'F59E12.16a']\n",
      "DANRE@7955@QfO22\n",
      "{'name': 'Danio rerio', 'uniprot': 'UP000000437', 'source': 'RefSeq', 'refseq': 'GCF_000002035.6'}\n",
      "Gene_count 1156, No overlap 2, Cant map set 0, Multi map set 1\n",
      "Multi map set\n",
      "{'NP_001017751.2'}\n",
      "Cant map examples\n",
      "set()\n",
      "DROME@7227@QfO22\n",
      "{'name': 'Drosophila melanogaster', 'uniprot': 'UP000000803', 'source': 'ENA/EMBL', 'refseq': 'GCA_000001215.4'}\n",
      "Gene_count 961, No overlap 0, Cant map set 6, Multi map set 1\n",
      "Multi map set\n",
      "{'AAM68345.2'}\n",
      "Cant map examples\n",
      "['AAF45822.1', 'AAF48554.1', 'AAN12244.1']\n",
      "GALGA@9031@000002315_5\n",
      "{'name': 'Gallus gallus', 'uniprot': 'UP000000539', 'source': 'Ensembl', 'refseq': 'GCA_000002315.5'}\n",
      "Gene_count 917, No overlap 8, Cant map set 22, Multi map set 2\n",
      "Multi map set\n",
      "{'ENSGALG00000055892', 'ENSGALG00000058203'}\n",
      "Cant map examples\n",
      "['ENSGALG00000058230', 'ENSGALG00000056352', 'ENSGALG00000055477']\n",
      "HELRO@6412@QfO22\n",
      "{'name': 'Helobdella robusta', 'uniprot': 'UP000015101', 'source': 'Ensembl', 'refseq': 'GCA_000326865.1'}\n",
      "Gene_count 972, No overlap 64, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "IXOSC@6945@QfO22\n",
      "{'name': 'Ixodes scapularis', 'uniprot': 'UP000001555', 'source': 'Ensembl', 'refseq': 'GCA_000208615.1'}\n",
      "Gene_count 919, No overlap 66, Cant map set 16, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['EEC03934.1', 'EEC10570.1', 'EEC20238.1']\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 957, No overlap 20, Cant map set 35, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['NEMVEDRAFT_v1g116756', 'NEMVEDRAFT_v1g2076', 'NEMVEDRAFT_v1g91541']\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 1276, No overlap 169, Cant map set 2, Multi map set 12\n",
      "Multi map set\n",
      "{'ENSRNOG00000040040', 'ENSRNOG00000030447', '621291', 'ENSRNOP00000016399', '2431', '1303019', 'ENSRNOG00000004477', 'ENSRNOP00000057879', 'ENSRNOP00000061132', '620870', 'ENSRNOG00000010587', 'Rnf10'}\n",
      "Cant map examples\n",
      "{'ENSRNOP00000051188', 'ENSRNOG00000032348'}\n",
      "TRICA@7070@QfO22\n",
      "{'name': 'Tribolium castaneum', 'uniprot': 'UP000007266', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002335.3'}\n",
      "Gene_count 960, No overlap 4, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "XENTR@8364@QfO22\n",
      "{'name': 'Xenopus tropicalis', 'uniprot': 'UP000008143', 'source': 'RefSeq', 'refseq': 'GCF_000004195.4'}\n",
      "Gene_count 958, No overlap 4, Cant map set 0, Multi map set 47\n",
      "Multi map set\n",
      "{'poldip2', 'XP_012814396.2', 'XP_012814322.1', 'XP_012820148.2', 'rrn3', 'XP_012809948.1', 'XP_017951362.1', 'XP_012822075.1', 'bbs4', 'XP_017947959.1', 'XP_012814325.1', 'ift52', 'tmem209', 'XP_002938755.1', 'pwp1', 'XP_002931733.1', 'tfb1m', 'XP_002939365.2', 'NP_001016895.1', 'XP_012814323.1', 'mrps5', 'XP_017950881.1', 'NP_001027490.1', 'pus1', 'NP_001165133.1', 'NP_001121527.1', 'XP_004914765.1', 'rpl28', 'med13', 'mrpl47', 'XP_012822063.1', 'nup107', 'utp18', 'tm2d3', 'XP_002934120.1', 'c12orf4', 'ctu2', 'XP_012814324.1', 'XP_002936868.2', 'tpd52', 'det1', 'supt4h1', 'XP_004912996.1', 'NP_989259.2', 'XP_004912577.1', 'ankmy2', 'tmem70'}\n",
      "Cant map examples\n",
      "set()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6239</td>\n",
       "      <td>739</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7955</td>\n",
       "      <td>1156</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7227</td>\n",
       "      <td>961</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9031</td>\n",
       "      <td>917</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6412</td>\n",
       "      <td>972</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6945</td>\n",
       "      <td>919</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>45351</td>\n",
       "      <td>957</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10116</td>\n",
       "      <td>1276</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7070</td>\n",
       "      <td>960</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8364</td>\n",
       "      <td>958</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0    6239         739           1\n",
       "1    7955        1156           2\n",
       "2    7227         961           0\n",
       "3    9031         917           8\n",
       "4    6412         972          64\n",
       "5    6945         919          66\n",
       "6   45351         957          20\n",
       "7   10116        1276         169\n",
       "8    7070         960           4\n",
       "9    8364         958           4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(\"BUSCO metaeuk\")\n",
    "busco_metaeuk_df = get_overlap_table(busco_metaeuk_pkl)\n",
    "#busco_metaeuk_df.to_csv('../overlap_tables/busco_metaeuk_overlap_gff_files.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "59318ede-a322-4c9f-ab13-69391773525f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### fDOG-Assembly Augustus ########\n",
      "GALGA@9031@000002315_5\n",
      "{'name': 'Gallus gallus', 'uniprot': 'UP000000539', 'source': 'Ensembl', 'refseq': 'GCA_000002315.5'}\n",
      "Gene_count 897, No overlap 7, Cant map set 20, Multi map set 2\n",
      "Multi map set\n",
      "{'ENSGALP00000078034', 'TRAPPC2L'}\n",
      "Cant map examples\n",
      "['ENSGALG00000059040', 'ENSGALP00000091600', 'ENSGALG00000059518']\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 1009, No overlap 45, Cant map set 1, Multi map set 9\n",
      "Multi map set\n",
      "{'Ublcp1', '1561342', 'Rnf10', '620507', 'ENSRNOP00000058286', 'ENSRNOG00000055257', 'ENSRNOG00000013223', 'ENSRNOG00000002241', 'Ak6'}\n",
      "Cant map examples\n",
      "{'ENSRNOP00000049054'}\n",
      "TRICA@7070@QfO22\n",
      "{'name': 'Tribolium castaneum', 'uniprot': 'UP000007266', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002335.3'}\n",
      "Gene_count 945, No overlap 6, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "CAEEL@6239@QfO22\n",
      "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n",
      "Gene_count 728, No overlap 3, Cant map set 6, Multi map set 1\n",
      "Multi map set\n",
      "{'CELE_Y39A3CR.1'}\n",
      "Cant map examples\n",
      "['CELE_F09E5.9', 'cds-CELE_Y57G11C.21', 'CELE_T08D2.2']\n",
      "DROME@7227@QfO22\n",
      "{'name': 'Drosophila melanogaster', 'uniprot': 'UP000000803', 'source': 'ENA/EMBL', 'refseq': 'GCA_000001215.4'}\n",
      "Gene_count 926, No overlap 6, Cant map set 5, Multi map set 1\n",
      "Multi map set\n",
      "{'Dmel_CG7843'}\n",
      "Cant map examples\n",
      "['Dmel_CG9742', 'AAN12244.1', 'AAF48554.1']\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 872, No overlap 32, Cant map set 39, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['NEMVEDRAFT_v1g88439', 'NEMVEDRAFT_v1g116756', 'NEMVEDRAFT_v1g10259']\n",
      "IXOSC@6945@QfO22\n",
      "{'name': 'Ixodes scapularis', 'uniprot': 'UP000001555', 'source': 'Ensembl', 'refseq': 'GCA_000208615.1'}\n",
      "Gene_count 863, No overlap 107, Cant map set 11, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['IscW_ISCW012441', 'IscW_ISCW018975', 'IscW_ISCW007765']\n",
      "DANRE@7955@QfO22\n",
      "{'name': 'Danio rerio', 'uniprot': 'UP000000437', 'source': 'RefSeq', 'refseq': 'GCF_000002035.6'}\n",
      "Gene_count 1164, No overlap 6, Cant map set 0, Multi map set 1\n",
      "Multi map set\n",
      "{'NP_001017751.2'}\n",
      "Cant map examples\n",
      "set()\n",
      "HELRO@6412@QfO22\n",
      "{'name': 'Helobdella robusta', 'uniprot': 'UP000015101', 'source': 'Ensembl', 'refseq': 'GCA_000326865.1'}\n",
      "Gene_count 803, No overlap 26, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "XENTR@8364@QfO22\n",
      "{'name': 'Xenopus tropicalis', 'uniprot': 'UP000008143', 'source': 'RefSeq', 'refseq': 'GCF_000004195.4'}\n",
      "Gene_count 949, No overlap 5, Cant map set 0, Multi map set 45\n",
      "Multi map set\n",
      "{'XP_004912577.1', 'NP_001016895.1', 'NP_001107535.2', 'supt4h1', 'XP_004912998.1', 'slc35c1', 'det1', 'XP_002937119.1', 'elp4', 'fdxr', 'sympk', 'XP_017950881.1', 'XP_012814321.1', 'gtf2h1', 'XP_012822075.1', 'ctu2', 'XP_004912995.1', 'XP_002937728.1', 'NP_001016714.1', 'XP_004912996.1', 'pwp1', 'XP_012814322.1', 'vps54', 'NP_001016635.1', 'NP_001017098.1', 'c1qbp', 'XP_012822063.1', 'nup107', 'NP_001007514.1', 'ift52', 'XP_017951362.1', 'XP_017949732.1', 'XP_012820022.1', 'XP_012814323.1', 'NP_001027490.1', 'bbs4', 'XP_002937906.1', 'XP_012820148.2', 'XP_002933349.1', 'NP_001165133.1', 'XP_004918010.1', 'XP_012814324.1', 'XP_012814325.1', 'tmem209', 'pus1'}\n",
      "Cant map examples\n",
      "set()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9031</td>\n",
       "      <td>897</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10116</td>\n",
       "      <td>1009</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7070</td>\n",
       "      <td>945</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6239</td>\n",
       "      <td>728</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7227</td>\n",
       "      <td>926</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>45351</td>\n",
       "      <td>872</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6945</td>\n",
       "      <td>863</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7955</td>\n",
       "      <td>1164</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6412</td>\n",
       "      <td>803</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8364</td>\n",
       "      <td>949</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0    9031         897           7\n",
       "1   10116        1009          45\n",
       "2    7070         945           6\n",
       "3    6239         728           3\n",
       "4    7227         926           6\n",
       "5   45351         872          32\n",
       "6    6945         863         107\n",
       "7    7955        1164           6\n",
       "8    6412         803          26\n",
       "9    8364         949           5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('##### fDOG-Assembly Augustus ########')\n",
    "fdog_busco_augustus_df = get_overlap_table(fdog_augustus_pkl)\n",
    "#fdog_busco_augustus_df.to_csv('../overlap_tables/fdog_ass_busco_augustus_overlap_gff_files_gallus_v2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7046dbea-50f0-4044-86cb-cdf7b70a6cf0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### fDOG-Assembly MetaEuk ########\n",
      "CAEEL@6239@QfO22\n",
      "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n",
      "Gene_count 730, No overlap 2, Cant map set 9, Multi map set 1\n",
      "Multi map set\n",
      "{'CELE_Y39A3CR.1'}\n",
      "Cant map examples\n",
      "['CELE_F52C6.13', 'F47G3.2', 'CELE_F55A3.7']\n",
      "DANRE@7955@QfO22\n",
      "{'name': 'Danio rerio', 'uniprot': 'UP000000437', 'source': 'RefSeq', 'refseq': 'GCF_000002035.6'}\n",
      "Gene_count 1159, No overlap 9, Cant map set 0, Multi map set 1\n",
      "Multi map set\n",
      "{'NP_001017751.2'}\n",
      "Cant map examples\n",
      "set()\n",
      "TRICA@7070@QfO22\n",
      "{'name': 'Tribolium castaneum', 'uniprot': 'UP000007266', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002335.3'}\n",
      "Gene_count 949, No overlap 11, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "DROME@7227@QfO22\n",
      "{'name': 'Drosophila melanogaster', 'uniprot': 'UP000000803', 'source': 'ENA/EMBL', 'refseq': 'GCA_000001215.4'}\n",
      "Gene_count 934, No overlap 10, Cant map set 10, Multi map set 1\n",
      "Multi map set\n",
      "{'Dmel_CG7843'}\n",
      "Cant map examples\n",
      "['AAN12244.1', 'UYK33078.1', 'AAF46297.1']\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 1039, No overlap 73, Cant map set 42, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['NEMVEDRAFT_v1g31564', 'cds-NEMVEDRAFT_v1g14357', 'NEMVEDRAFT_v1g41253']\n",
      "IXOSC@6945@QfO22\n",
      "{'name': 'Ixodes scapularis', 'uniprot': 'UP000001555', 'source': 'Ensembl', 'refseq': 'GCA_000208615.1'}\n",
      "Gene_count 967, No overlap 176, Cant map set 10, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['IscW_ISCW012893', 'IscW_ISCW007765', 'EEC20238.1']\n",
      "HELRO@6412@QfO22\n",
      "{'name': 'Helobdella robusta', 'uniprot': 'UP000015101', 'source': 'Ensembl', 'refseq': 'GCA_000326865.1'}\n",
      "Gene_count 871, No overlap 22, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "XENTR@8364@QfO22\n",
      "{'name': 'Xenopus tropicalis', 'uniprot': 'UP000008143', 'source': 'RefSeq', 'refseq': 'GCF_000004195.4'}\n",
      "Gene_count 974, No overlap 29, Cant map set 0, Multi map set 44\n",
      "Multi map set\n",
      "{'XP_004912577.1', 'NP_001016895.1', 'NP_001107535.2', 'supt4h1', 'XP_004912998.1', 'det1', 'slc35c1', 'XP_002937119.1', 'med13', 'elp4', 'XP_017950881.1', 'sympk', 'fdxr', 'XP_012814321.1', 'gtf2h1', 'XP_012822075.1', 'ctu2', 'XP_004912995.1', 'NP_001016714.1', 'XP_002937728.1', 'c1qbp', 'vps54', 'XP_012814322.1', 'XP_004912996.1', 'pwp1', 'XP_012822063.1', 'NP_001016635.1', 'nup107', 'NP_001007514.1', 'XP_017951362.1', 'ift52', 'XP_017949732.1', 'XP_012814323.1', 'NP_001027490.1', 'bbs4', 'XP_002937906.1', 'XP_012820148.2', 'XP_002933349.1', 'NP_001165133.1', 'XP_004918010.1', 'XP_012814324.1', 'XP_012814325.1', 'tmem209', 'pus1'}\n",
      "Cant map examples\n",
      "set()\n",
      "GALGA@9031@000002315_5\n",
      "{'name': 'Gallus gallus', 'uniprot': 'UP000000539', 'source': 'Ensembl', 'refseq': 'GCA_000002315.5'}\n",
      "Gene_count 921, No overlap 28, Cant map set 20, Multi map set 2\n",
      "Multi map set\n",
      "{'ENSGALP00000078034', 'TRAPPC2L'}\n",
      "Cant map examples\n",
      "['ENSGALP00000094649', 'ENSGALP00000090518', 'ENSGALP00000089625']\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 1071, No overlap 112, Cant map set 1, Multi map set 9\n",
      "Multi map set\n",
      "{'Ublcp1', '1561342', 'Rnf10', '620507', 'ENSRNOP00000058286', 'ENSRNOG00000055257', 'ENSRNOG00000013223', 'ENSRNOG00000002241', 'Ak6'}\n",
      "Cant map examples\n",
      "{'ENSRNOP00000049054'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6239</td>\n",
       "      <td>730</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7955</td>\n",
       "      <td>1159</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7070</td>\n",
       "      <td>949</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7227</td>\n",
       "      <td>934</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>45351</td>\n",
       "      <td>1039</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6945</td>\n",
       "      <td>967</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6412</td>\n",
       "      <td>871</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8364</td>\n",
       "      <td>974</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9031</td>\n",
       "      <td>921</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10116</td>\n",
       "      <td>1071</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0    6239         730           2\n",
       "1    7955        1159           9\n",
       "2    7070         949          11\n",
       "3    7227         934          10\n",
       "4   45351        1039          73\n",
       "5    6945         967         176\n",
       "6    6412         871          22\n",
       "7    8364         974          29\n",
       "8    9031         921          28\n",
       "9   10116        1071         112"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('##### fDOG-Assembly MetaEuk ########')\n",
    "fdog_busco_metaeuk_df = get_overlap_table(fdog_metaeuk_pkl)\n",
    "#fdog_busco_metaeuk_df.to_csv('../overlap_tables/fdog_ass_busco_metaeuk_overlap_gff_files_gallus_v2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "392c4e76-97b5-4ccc-9fca-b9224acda8ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84a669c0-313b-44c6-bcfa-d5b67913240b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4737bca6-f3a5-49b8-92ff-65cbcc83532a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### fDOG-Assembly MetaEuk sens########\n",
      "CAEEL@6239@QfO22\n",
      "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n",
      "Gene_count 737, No overlap 11, Cant map set 8, Multi map set 2\n",
      "Multi map set\n",
      "{'Y39A3CR.1c', 'CELE_Y39A3CR.1'}\n",
      "Cant map examples\n",
      "['CELE_Y102A5C.6', 'cds-CELE_F52C6.13', 'cds-CELE_F47G3.2']\n",
      "TRICA@7070@QfO22\n",
      "{'name': 'Tribolium castaneum', 'uniprot': 'UP000007266', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002335.3'}\n",
      "Gene_count 955, No overlap 16, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "DROME@7227@QfO22\n",
      "{'name': 'Drosophila melanogaster', 'uniprot': 'UP000000803', 'source': 'ENA/EMBL', 'refseq': 'GCA_000001215.4'}\n",
      "Gene_count 944, No overlap 16, Cant map set 7, Multi map set 1\n",
      "Multi map set\n",
      "{'Dmel_CG7843'}\n",
      "Cant map examples\n",
      "['AAF48634.1', 'AAF48554.1', 'Dmel_CG46509']\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 1048, No overlap 81, Cant map set 43, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['cds-NEMVEDRAFT_v1g86020', 'NEMVEDRAFT_v1g24420', 'NEMVEDRAFT_v1g66999']\n",
      "DANRE@7955@QfO22\n",
      "{'name': 'Danio rerio', 'uniprot': 'UP000000437', 'source': 'RefSeq', 'refseq': 'GCF_000002035.6'}\n",
      "Gene_count 1172, No overlap 10, Cant map set 0, Multi map set 1\n",
      "Multi map set\n",
      "{'NP_001017751.2'}\n",
      "Cant map examples\n",
      "set()\n",
      "GALGA@9031@000002315_5\n",
      "{'name': 'Gallus gallus', 'uniprot': 'UP000000539', 'source': 'Ensembl', 'refseq': 'GCA_000002315.5'}\n",
      "Gene_count 926, No overlap 39, Cant map set 18, Multi map set 2\n",
      "Multi map set\n",
      "{'ENSGALG00000055892', 'ENSGALP00000078034'}\n",
      "Cant map examples\n",
      "['ENSGALG00000058652', 'THUMPD1', 'ENSGALP00000086144']\n",
      "HELRO@6412@QfO22\n",
      "{'name': 'Helobdella robusta', 'uniprot': 'UP000015101', 'source': 'Ensembl', 'refseq': 'GCA_000326865.1'}\n",
      "Gene_count 871, No overlap 28, Cant map set 0, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "set()\n",
      "XENTR@8364@QfO22\n",
      "{'name': 'Xenopus tropicalis', 'uniprot': 'UP000008143', 'source': 'RefSeq', 'refseq': 'GCF_000004195.4'}\n",
      "Gene_count 981, No overlap 39, Cant map set 0, Multi map set 42\n",
      "Multi map set\n",
      "{'NP_001016895.1', 'XP_017947958.1', 'nup107', 'XP_002939365.2', 'NP_001016679.1', 'XP_004914765.1', 'ctu2', 'NP_001107535.2', 'XP_017947959.1', 'NP_001165133.1', 'ubiad1', 'XP_004912705.1', 'tmem209', 'c12orf4', 'pwp1', 'c1qbp', 'XP_004918010.1', 'XP_012827224.2', 'hsd17b4', 'XP_002937728.1', 'pus1', 'elp4', 'XP_012814322.1', 'XP_012820148.2', 'XP_012809948.1', 'XP_017949732.1', 'NP_001016494.1', 'fdxr', 'XP_002934120.1', 'XP_017951362.1', 'sympk', 'XP_004912577.1', 'det1', 'XP_002937906.1', 'XP_004911679.1', 'ift52', 'NP_001016635.1', 'tmem70', 'rpl28', 'bbs4', 'XP_012822075.1', 'ankmy2'}\n",
      "Cant map examples\n",
      "set()\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 1080, No overlap 108, Cant map set 1, Multi map set 9\n",
      "Multi map set\n",
      "{'1309085', '620507', 'ENSRNOP00000063855', '7387335', 'ENSRNOP00000016399', 'ENSRNOG00000004477', 'Hspbap1', 'ENSRNOP00000061132', 'ENSRNOP00000075579'}\n",
      "Cant map examples\n",
      "{'ENSRNOG00000032348'}\n",
      "IXOSC@6945@QfO22\n",
      "{'name': 'Ixodes scapularis', 'uniprot': 'UP000001555', 'source': 'Ensembl', 'refseq': 'GCA_000208615.1'}\n",
      "Gene_count 972, No overlap 187, Cant map set 12, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['EEC01038.1', 'IscW_ISCW014590', 'IscW_ISCW015320']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6239</td>\n",
       "      <td>737</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7070</td>\n",
       "      <td>955</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7227</td>\n",
       "      <td>944</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>45351</td>\n",
       "      <td>1048</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7955</td>\n",
       "      <td>1172</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9031</td>\n",
       "      <td>926</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6412</td>\n",
       "      <td>871</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8364</td>\n",
       "      <td>981</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10116</td>\n",
       "      <td>1080</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6945</td>\n",
       "      <td>972</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0    6239         737          11\n",
       "1    7070         955          16\n",
       "2    7227         944          16\n",
       "3   45351        1048          81\n",
       "4    7955        1172          10\n",
       "5    9031         926          39\n",
       "6    6412         871          28\n",
       "7    8364         981          39\n",
       "8   10116        1080         108\n",
       "9    6945         972         187"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('##### fDOG-Assembly MetaEuk sens########')\n",
    "fdog_busco_metaeuk_df = get_overlap_table(fdog_metaeuk_sens_pkl)\n",
    "fdog_busco_metaeuk_df.to_csv('../overlap_tables/fdog_ass_busco_metaeuk_overlap_gff_files_gallus_v2_sens.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf9c203-3ec9-440a-88c1-8e405f1b63cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### output fDOG-Assembly augustus, new chicken, always iterating whole contig\n",
    "RATNO@10116@QfO22\n",
    "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
    "Gene_count 1008, No overlap 45, Cant map 1, Multi map 9\n",
    "Multi map set\n",
    "{'1561342', 'ENSRNOG00000063018', '61932', 'ENSRNOP00000075579', '620870', 'Rnf10', 'ENSRNOP00000014632', 'ENSRNOG00000004477', 'ENSRNOP00000058286'}\n",
    "Cant map examples\n",
    "{'ENSRNOP00000049054'}\n",
    "GALGA@9031@000002315_5\n",
    "{'name': 'Gallus gallus', 'uniprot': 'UP000000539', 'source': 'Ensembl', 'refseq': 'GCA_000002315.5'}\n",
    "Gene_count 898, No overlap 7, Cant map 20, Multi map 2\n",
    "Multi map set\n",
    "{'ENSGALP00000074383', 'ENSGALP00000090837'}\n",
    "Cant map examples\n",
    "['ENSGALP00000076009', 'ENSGALP00000078129', 'THUMPD1']\n",
    "TRICA@7070@QfO22\n",
    "{'name': 'Tribolium castaneum', 'uniprot': 'UP000007266', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002335.3'}\n",
    "Gene_count 946, No overlap 6, Cant map 0, Multi map 0\n",
    "Multi map set\n",
    "set()\n",
    "Cant map examples\n",
    "set()\n",
    "DROME@7227@QfO22\n",
    "{'name': 'Drosophila melanogaster', 'uniprot': 'UP000000803', 'source': 'ENA/EMBL', 'refseq': 'GCA_000001215.4'}\n",
    "Gene_count 927, No overlap 6, Cant map 5, Multi map 1\n",
    "Multi map set\n",
    "{'AAM68345.2'}\n",
    "Cant map examples\n",
    "['Dmel_CG9177', 'Dmel_CG1524', 'Dmel_CG9742']\n",
    "DANRE@7955@QfO22\n",
    "{'name': 'Danio rerio', 'uniprot': 'UP000000437', 'source': 'RefSeq', 'refseq': 'GCF_000002035.6'}\n",
    "Gene_count 1163, No overlap 6, Cant map 0, Multi map 1\n",
    "Multi map set\n",
    "{'osgep'}\n",
    "Cant map examples\n",
    "set()\n",
    "IXOSC@6945@QfO22\n",
    "{'name': 'Ixodes scapularis', 'uniprot': 'UP000001555', 'source': 'Ensembl', 'refseq': 'GCA_000208615.1'}\n",
    "Gene_count 863, No overlap 107, Cant map 11, Multi map 0\n",
    "Multi map set\n",
    "set()\n",
    "Cant map examples\n",
    "['IscW_ISCW018510', 'EEC16001.1', 'IscW_ISCW018975']\n",
    "GALGA@9031@QfO22\n",
    "XENTR@8364@QfO22\n",
    "{'name': 'Xenopus tropicalis', 'uniprot': 'UP000008143', 'source': 'RefSeq', 'refseq': 'GCF_000004195.4'}\n",
    "Gene_count 949, No overlap 5, Cant map 0, Multi map 47\n",
    "Multi map set\n",
    "{'mrps5', 'XP_004914765.1', 'rrn3', 'XP_004913000.1', 'XP_012814324.1', 'c1qbp', 'XP_002938755.1', 'XP_002937906.1', 'supt4h1', 'NP_001007514.1', 'tbc1d7', 'XP_012814323.1', 'NP_001011344.1', 'XP_004918010.1', 'XP_012809948.1', 'XP_004912881.1', 'rpl28', 'nup107', 'gbf1', 'NP_001016280.1', 'XP_002934120.1', 'ctu2', 'c12orf4', 'NP_001016587.1', 'XP_012814321.1', 'XP_002939365.2', 'XP_012822075.1', 'XP_012814396.2', 'ankmy2', 'XP_004912577.1', 'XP_017947958.1', 'NP_989259.2', 'XP_002938554.1', 'NP_001165133.1', 'det1', 'NP_001016895.1', 'mrpl47', 'XP_012814322.1', 'fdxr', 'mst1r', 'tmem70', 'utp18', 'NP_001027490.1', 'XP_004912704.1', 'poldip2', 'pus1', 'ubiad1'}\n",
    "Cant map examples\n",
    "set()\n",
    "HELRO@6412@QfO22\n",
    "{'name': 'Helobdella robusta', 'uniprot': 'UP000015101', 'source': 'Ensembl', 'refseq': 'GCA_000326865.1'}\n",
    "Gene_count 803, No overlap 26, Cant map 0, Multi map 0\n",
    "Multi map set\n",
    "set()\n",
    "Cant map examples\n",
    "set()\n",
    "NEMVE@45351@QfO22\n",
    "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
    "Gene_count 873, No overlap 32, Cant map 39, Multi map 0\n",
    "Multi map set\n",
    "set()\n",
    "Cant map examples\n",
    "['cds-NEMVEDRAFT_v1g41253', 'NEMVEDRAFT_v1g2076', 'NEMVEDRAFT_v1g89647']\n",
    "CAEEL@6239@QfO22\n",
    "{'name': 'Caenorhabditis elegans', 'uniprot': 'UP000001940', 'source': 'ENA/EMBL', 'refseq': 'GCA_000002985.3'}\n",
    "Gene_count 727, No overlap 3, Cant map 6, Multi map 3\n",
    "Multi map set\n",
    "{'Y39A3CR.1d', 'Y39A3CR.1b', 'Y39A3CR.1c'}\n",
    "Cant map examples\n",
    "['Y57G11C.21', 'cds-CELE_F47G3.2', 'F55A3.7']\n",
    "14300 14300 14300 14300 14300 14300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2a26f55c-53dc-4d07-8cee-feaa0945b8cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[131423669, 131423695, '.', '-', '0', {'Atp6v0b', 'ENSRNOP00000026704', 'ENSRNOG00000019655'}, 'ENSRNOT00000026704'], [131424277, 131424467, '.', '-', '2', {'Atp6v0b', 'ENSRNOP00000026704', 'ENSRNOG00000019655'}, 'ENSRNOT00000026704'], [131424738, 131424789, '.', '-', '0', {'Atp6v0b', 'ENSRNOP00000026704', 'ENSRNOG00000019655'}, 'ENSRNOT00000026704'], [131424914, 131424983, '.', '-', '1', {'Atp6v0b', 'ENSRNOP00000026704', 'ENSRNOG00000019655'}, 'ENSRNOT00000026704'], [131425179, 131425256, '.', '-', '1', {'Atp6v0b', 'ENSRNOP00000026704', 'ENSRNOG00000019655'}, 'ENSRNOT00000026704'], [131425386, 131425444, '.', '-', '0', {'Atp6v0b', 'ENSRNOP00000026704', 'ENSRNOG00000019655'}, 'ENSRNOT00000026704']]\n"
     ]
    }
   ],
   "source": [
    "print(positions_gff['10116']['5']['ENSRNOG00000019655'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "46a33fbe-a6b9-4b8d-b56e-d12eea558fe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[10661863, 10661875, '.', '+', '0', {'rpl23', 'NP_001011231.1'}, 1], [10663371, 10663454, '.', '+', '2', {'rpl23', 'NP_001011231.1'}, 1], [10665060, 10665188, '.', '+', '2', {'rpl23', 'NP_001011231.1'}, 1], [10665520, 10665633, '.', '+', '2', {'rpl23', 'NP_001011231.1'}, 1], [10667663, 10667745, '.', '+', '2', {'rpl23', 'NP_001011231.1'}, 1]]\n"
     ]
    }
   ],
   "source": [
    "print(positions_gff['8364']['NC_030686.2']['rpl23'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "974f58c4-19e5-4b0f-a9ed-fe69ac73dbde",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'537470at33208_CM026978_1_1_g4.t1': [['AUGUSTUS', 'CDS', 131423669, 131423695, '-', '0', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1'], ['AUGUSTUS', 'CDS', 131424277, 131424467, '-', '2', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1'], ['AUGUSTUS', 'CDS', 131424738, 131424789, '-', '0', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1'], ['AUGUSTUS', 'CDS', 131424914, 131424983, '-', '1', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1'], ['AUGUSTUS', 'CDS', 131425179, 131425256, '-', '1', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1'], ['AUGUSTUS', 'CDS', 131425386, 131425469, '-', '1', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1'], ['AUGUSTUS', 'CDS', 131425680, 131425728, '-', '2', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1'], ['AUGUSTUS', 'CDS', 131426244, 131426310, '-', '0', 'ID=537470at33208_CM026978_1_1_g4.t1.cds;Parent=537470at33208_CM026978_1_1_g4.t1']]}\n"
     ]
    }
   ],
   "source": [
    "print(fdog_augustus_pkl['RATNO@10116@QfO22']['CM026978.1']['537470at33208'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15a0de01-0888-4d81-bcb2-e074f5c1e81b",
   "metadata": {},
   "source": [
    "## Human proteom fDOG-Assembly augustus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "81f61d35-5b57-4e4b-99fe-b5ed892abf45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "RATNO@10116@QfO22\n",
      "AY172581.1\n",
      "10116\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 24025, No overlap 2668, Cant map 10, Multi map 130\n",
      "Multi map set\n",
      "{'ENSRNOP00000048081', '1303057', '1303008', 'ENSRNOG00000016029', 'Eci2', 'ENSRNOG00000003846', '2197', 'Il17f', '621512', 'ENSRNOG00000042111', '1303049', 'ENSRNOG00000033747', 'Ggt1', '2497', 'ENSRNOG00000024243', 'ENSRNOP00000059218', 'Ppp1r2', 'ENSRNOG00000018650', '1307768', 'Slc30a8', '2504', '621704', 'ENSRNOP00000024406', 'ENSRNOG00000003259', '1306313', 'ENSRNOG00000023433', 'Fev', '620371', 'ENSRNOG00000005592', 'ENSRNOG00000013223', 'ENSRNOP00000001449', 'Ncmap', 'ENSRNOP00000068417', 'Taco1', 'ENSRNOP00000006573', 'Hand2', '735199', 'Tor2a', '620438', 'ENSRNOP00000074503', 'ENSRNOP00000072729', 'ENSRNOG00000032517', 'Selenop', 'ENSRNOP00000071934', 'ENSRNOP00000016399', 'ENSRNOP00000012222', 'Lrit1', 'ENSRNOG00000000900', 'Rnf10', 'Camkmt', 'ENSRNOG00000025365', 'LOC100364191', 'ENSRNOP00000012365', '61900', 'ENSRNOP00000073010', 'ENSRNOP00000074809', '1310386', 'ENSRNOG00000001720', '1303193', 'ENSRNOP00000065961', '1562672', 'ENSRNOG00000008981', 'ENSRNOG00000060237', 'Txnrd2', 'ENSRNOP00000073497', 'ENSRNOG00000008939', 'ENSRNOG00000004989', 'ENSRNOP00000066577', 'Selenof', 'ENSRNOP00000075824', 'ENSRNOG00000042152', 'ENSRNOP00000008705', 'ENSRNOG00000014187', 'ENSRNOP00000068116', '620647', '1306802', 'Psmf1', 'ENSRNOG00000004870', 'Mme', 'ENSRNOP00000066631', 'Htra4', 'ENSRNOP00000040016', 'ENSRNOG00000053181', 'ENSRNOP00000023685', 'Stra6', 'ENSRNOG00000029574', 'Ndufaf6', 'ENSRNOG00000049882', 'Crnkl1', 'Apoa4', 'Ak6', 'Cnksr3', 'ENSRNOP00000000383', '621129', '1303129', '708365', 'ENSRNOG00000023548', 'ENSRNOG00000013389', 'Krt5', 'ENSRNOP00000017101', '628894', 'ENSRNOG00000015430', 'ENSRNOP00000072234', '1590928', 'Uvssa', 'Minar1', 'Gpx6', 'ENSRNOP00000015540', 'ENSRNOG00000001416', 'ENSRNOG00000021555', 'ENSRNOP00000068190', '1309698', 'ENSRNOG00000004146', 'Tmem168', 'ENSRNOP00000060718', '727971', 'ENSRNOP00000027615', 'ENSRNOP00000021625', '1564570', '2320289', 'C19h16orf46', '708548', 'Selenot', 'ENSRNOP00000070410', 'ENSRNOG00000013720', '628836', 'ENSRNOG00000020833', 'ENSRNOP00000007018', '620870', 'ENSRNOP00000068607'}\n",
      "Cant map examples\n",
      "['11455920', 'ENSRNOG00000032348', 'Gng5-ps5']\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 13392, No overlap 1178, Cant map 405, Multi map 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['NEMVEDRAFT_v1g86', 'cds-NEMVEDRAFT_v1g838', 'NEMVEDRAFT_v1g60721']\n",
      "XENTR@8364@QfO22\n",
      "Not right key\n",
      "XENTR@8364@QfO22\n",
      "GALGA@9031@QfO22\n",
      "Not right key\n",
      "GALGA@9031@QfO22\n",
      "CAEEL@6239@QfO22\n",
      "Not right key\n",
      "CAEEL@6239@QfO22\n",
      "DANRE@7955@QfO22\n",
      "Not right key\n",
      "DANRE@7955@QfO22\n",
      "TRICA@7070@QfO22\n",
      "Not right key\n",
      "TRICA@7070@QfO22\n",
      "HELRO@6412@QfO22\n",
      "Not right key\n",
      "HELRO@6412@QfO22\n",
      "DROME@7227@QfO22\n",
      "Not right key\n",
      "DROME@7227@QfO22\n",
      "IXOSC@6945@QfO22\n",
      "Not right key\n",
      "IXOSC@6945@QfO22\n",
      "67272 67272 67272 67272 67272 67272\n"
     ]
    }
   ],
   "source": [
    "#Augustus with mapping file for augustus ref species, all proteins\n",
    "fdog_human_proteom_augustus_pkl = open_pkl('../pkl_files/fdog_assembly_augustus_human_proteom_busco_CDS_positions.pkl')\n",
    "fdog_human_proteom_augustus_df = get_overlap_table(fdog_human_proteom_augustus_pkl)\n",
    "fdog_human_proteom_augustus_df.to_csv('../overlap_tables/fdog_ass_human_proteom_augustus_overlap_gff_files_rat_nema.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8f58132d-cfe2-4834-abf8-f9502713b82d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TRICA@7070@QfO22\n",
      "Not right key\n",
      "TRICA@7070@QfO22\n",
      "DROME@7227@QfO22\n",
      "Not right key\n",
      "DROME@7227@QfO22\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 3246, No overlap 262, Cant map set 158, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['cds-NEMVEDRAFT_v1g61375', 'cds-NEMVEDRAFT_v1g61676', 'NEMVEDRAFT_v1g12630']\n",
      "IXOSC@6945@QfO22\n",
      "Not right key\n",
      "IXOSC@6945@QfO22\n",
      "DANRE@7955@QfO22\n",
      "Not right key\n",
      "DANRE@7955@QfO22\n",
      "HELRO@6412@QfO22\n",
      "Not right key\n",
      "HELRO@6412@QfO22\n",
      "GALGA@9031@QfO22\n",
      "XENTR@8364@QfO22\n",
      "Not right key\n",
      "XENTR@8364@QfO22\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 5749, No overlap 664, Cant map set 3, Multi map set 26\n",
      "Multi map set\n",
      "{'ENSRNOG00000016429', 'Slc30a8', 'ENSRNOP00000048081', 'Fbp1', 'Selenop', 'ENSRNOP00000065923', 'Xkr6', '61932', 'ENSRNOP00000007455', 'ENSRNOG00000018650', 'ENSRNOG00000009640', 'Gdf10', '1310453', 'Rps4x-ps9', '628789', 'ENSRNOG00000049334', 'Arx', 'Gpr3', 'Spem2', 'ENSRNOP00000059218', '1592346', 'ENSRNOP00000071934', 'Rb1', '2913', 'Msh5', 'Aebp1'}\n",
      "Cant map examples\n",
      "{'41253011', 'ENSRNOG00000032605', 'ENSRNOG00000032348'}\n",
      "CAEEL@6239@QfO22\n",
      "Not right key\n",
      "CAEEL@6239@QfO22\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45351</td>\n",
       "      <td>3246</td>\n",
       "      <td>262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10116</td>\n",
       "      <td>5749</td>\n",
       "      <td>664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0   45351        3246         262\n",
       "1   10116        5749         664"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Augustus with mapping file for augustus ref species, 5000 human seed genes\n",
    "fdog_human_proteom_augustus_pkl = open_pkl('../pkl_files/fdog_assembly_augustus_human_proteom_5t_CDS_positions.pkl')\n",
    "fdog_human_proteom_augustus_df = get_overlap_table(fdog_human_proteom_augustus_pkl)\n",
    "fdog_human_proteom_augustus_df.to_csv('../overlap_tables/fdog_ass_human_proteom_augustus_overlap_gff_files_rat_nema_5t.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0d38ff2a-e096-4e81-8e0a-c5146c40e32f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HELRO@6412@QfO22\n",
      "Not right key\n",
      "HELRO@6412@QfO22\n",
      "GALGA@9031@QfO22\n",
      "XENTR@8364@QfO22\n",
      "Not right key\n",
      "XENTR@8364@QfO22\n",
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "RATNO@10116@QfO22\n",
      "AY172581.1\n",
      "10116\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 5647, No overlap 732, Cant map set 2, Multi map set 27\n",
      "Multi map set\n",
      "{'Slc30a8', 'ENSRNOP00000048081', 'Fbp1', 'Selenop', 'ENSRNOP00000065923', 'Xkr6', 'Cyp4f6', '61932', 'ENSRNOP00000007455', 'ENSRNOG00000018650', 'Syn3', 'Gdf10', 'ENSRNOG00000009640', '1310453', 'Rps4x-ps9', '628789', 'ENSRNOG00000049334', 'Arx', 'Gpr3', 'ENSRNOP00000059218', '1592346', 'Spem2', 'ENSRNOP00000071934', 'Rb1', '2913', 'Msh5', 'Aebp1'}\n",
      "Cant map examples\n",
      "{'41253011', 'ENSRNOG00000032348'}\n",
      "CAEEL@6239@QfO22\n",
      "Not right key\n",
      "CAEEL@6239@QfO22\n",
      "TRICA@7070@QfO22\n",
      "Not right key\n",
      "TRICA@7070@QfO22\n",
      "IXOSC@6945@QfO22\n",
      "Not right key\n",
      "IXOSC@6945@QfO22\n",
      "DROME@7227@QfO22\n",
      "Not right key\n",
      "DROME@7227@QfO22\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 4164, No overlap 643, Cant map set 137, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['NEMVEDRAFT_v1g100901', 'NEMVEDRAFT_v1g11044', 'NEMVEDRAFT_v1g40037']\n",
      "DANRE@7955@QfO22\n",
      "Not right key\n",
      "DANRE@7955@QfO22\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10116</td>\n",
       "      <td>5647</td>\n",
       "      <td>732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>45351</td>\n",
       "      <td>4164</td>\n",
       "      <td>643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0   10116        5647         732\n",
       "1   45351        4164         643"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#MetaEuk, 5000 human seed genes\n",
    "fdog_human_proteom_metaeuk_pkl = open_pkl('../pkl_files/fdog_assembly_metaeuk_human_proteom_5t_CDS_positions.pkl')\n",
    "fdog_human_proteom_metaeuk_df = get_overlap_table(fdog_human_proteom_metaeuk_pkl)\n",
    "fdog_human_proteom_metaeuk_df.to_csv('../overlap_tables/fdog_ass_human_proteom_metaeuk_overlap_gff_files_rat_nema_5t.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "282a39f8-3558-4483-a050-1db8d1bf0fbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RATNO@10116@QfO22\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "RATNO@10116@QfO22\n",
      "AY172581.1\n",
      "10116\n",
      "{'name': 'Rattus norvegicus', 'uniprot': 'UP000002494', 'source': 'Ensembl', 'refseq': 'GCA_015227675.2'}\n",
      "Gene_count 5689, No overlap 760, Cant map set 2, Multi map set 27\n",
      "Multi map set\n",
      "{'Slc30a8', 'ENSRNOP00000048081', 'Fbp1', 'Selenop', 'Xkr6', 'ENSRNOP00000065923', 'Cyp4f6', '61932', 'ENSRNOP00000007455', 'ENSRNOG00000018650', 'Syn3', 'ENSRNOG00000009640', 'Gdf10', '1310453', 'Rps4x-ps9', '628789', 'ENSRNOG00000049334', 'Arx', 'Gpr3', '1592346', 'Spem2', 'ENSRNOP00000059218', 'ENSRNOP00000071934', 'Rb1', '2913', 'Msh5', 'Aebp1'}\n",
      "Cant map examples\n",
      "{'41253011', 'ENSRNOG00000032348'}\n",
      "NEMVE@45351@QfO22\n",
      "{'name': 'Nematostella vectensis', 'uniprot': 'UP000001593', 'source': 'ENA/EMBL', 'refseq': 'GCA_000209225.1'}\n",
      "Gene_count 4235, No overlap 742, Cant map set 131, Multi map set 0\n",
      "Multi map set\n",
      "set()\n",
      "Cant map examples\n",
      "['cds-NEMVEDRAFT_v1g5601', 'NEMVEDRAFT_v1g25811', 'NEMVEDRAFT_v1g87917']\n",
      "IXOSC@6945@QfO22\n",
      "Not right key\n",
      "IXOSC@6945@QfO22\n",
      "CAEEL@6239@QfO22\n",
      "Not right key\n",
      "CAEEL@6239@QfO22\n",
      "DANRE@7955@QfO22\n",
      "Not right key\n",
      "DANRE@7955@QfO22\n",
      "DROME@7227@QfO22\n",
      "Not right key\n",
      "DROME@7227@QfO22\n",
      "TRICA@7070@QfO22\n",
      "Not right key\n",
      "TRICA@7070@QfO22\n",
      "HELRO@6412@QfO22\n",
      "Not right key\n",
      "HELRO@6412@QfO22\n",
      "GALGA@9031@QfO22\n",
      "XENTR@8364@QfO22\n",
      "Not right key\n",
      "XENTR@8364@QfO22\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>#Orthologs</th>\n",
       "      <th>No_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10116</td>\n",
       "      <td>5689</td>\n",
       "      <td>760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>45351</td>\n",
       "      <td>4235</td>\n",
       "      <td>742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species  #Orthologs  No_overlap\n",
       "0   10116        5689         760\n",
       "1   45351        4235         742"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#MetaEuk sens, 5000 human seed genes\n",
    "fdog_human_proteom_metaeuk_pkl = open_pkl('../pkl_files/fdog_assembly_metaeuk_sens_human_proteom_5t_CDS_positions.pkl')\n",
    "fdog_human_proteom_metaeuk_df = get_overlap_table(fdog_human_proteom_metaeuk_pkl)\n",
    "fdog_human_proteom_metaeuk_df.to_csv('../overlap_tables/fdog_ass_human_proteom_metaeuk_sens_overlap_gff_files_rat_nema_5t.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a323ae5-3204-4269-a84a-9e09c81eec03",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
